// @carbon/charts/dist/color-scale-utils-CiW9fERh.mjs
import { Z as Ne, s as $, V as Te, u as Hn, A as te, k as qn, K as ne, x as Un, p as zn, Q as Qn, D as Xn, S as ft, i as zt } from "./enums-DwsxZqjR.mjs";
import { easeCubicInOut as kn, pointer as Vn, min as Zn, extent as Kn, scaleLinear as Jn, scaleQuantize as er } from "d3";
function Ie(e) {
  return (t = {}) => {
    const n = t.width ? String(t.width) : e.defaultWidth;
    return e.formats[n] || e.formats[e.defaultWidth];
  };
}
function re(e) {
  return (t, n) => {
    const r = n != null && n.context ? String(n.context) : "standalone";
    let a;
    if (r === "formatting" && e.formattingValues) {
      const o = e.defaultFormattingWidth || e.defaultWidth, s = n != null && n.width ? String(n.width) : o;
      a = e.formattingValues[s] || e.formattingValues[o];
    } else {
      const o = e.defaultWidth, s = n != null && n.width ? String(n.width) : e.defaultWidth;
      a = e.values[s] || e.values[o];
    }
    const i = e.argumentCallback ? e.argumentCallback(t) : t;
    return a[i];
  };
}
function ae(e) {
  return (t, n = {}) => {
    const r = n.width, a = r && e.matchPatterns[r] || e.matchPatterns[e.defaultMatchWidth], i = t.match(a);
    if (!i)
      return null;
    const o = i[0], s = r && e.parsePatterns[r] || e.parsePatterns[e.defaultParseWidth], u = Array.isArray(s) ? nr(s, (f) => f.test(o)) : (
      // eslint-disable-next-line @typescript-eslint/no-explicit-any -- I challange you to fix the type
      tr(s, (f) => f.test(o))
    );
    let c;
    c = e.valueCallback ? e.valueCallback(u) : u, c = n.valueCallback ? (
      // eslint-disable-next-line @typescript-eslint/no-explicit-any -- I challange you to fix the type
      n.valueCallback(c)
    ) : c;
    const l = t.slice(o.length);
    return { value: c, rest: l };
  };
}
function tr(e, t) {
  for (const n in e)
    if (Object.prototype.hasOwnProperty.call(e, n) && t(e[n]))
      return n;
}
function nr(e, t) {
  for (let n = 0; n < e.length; n++)
    if (t(e[n]))
      return n;
}
function rr(e) {
  return (t, n = {}) => {
    const r = t.match(e.matchPattern);
    if (!r)
      return null;
    const a = r[0], i = t.match(e.parsePattern);
    if (!i)
      return null;
    let o = e.valueCallback ? e.valueCallback(i[0]) : i[0];
    o = n.valueCallback ? n.valueCallback(o) : o;
    const s = t.slice(a.length);
    return { value: o, rest: s };
  };
}
function A(e) {
  const t = Object.prototype.toString.call(e);
  return e instanceof Date || typeof e == "object" && t === "[object Date]" ? new e.constructor(+e) : typeof e == "number" || t === "[object Number]" || typeof e == "string" || t === "[object String]" ? new Date(e) : /* @__PURE__ */ new Date(NaN);
}
let ar = {};
function Pe() {
  return ar;
}
function ue(e, t) {
  var s, u, c, l;
  const n = Pe(), r = (t == null ? void 0 : t.weekStartsOn) ?? ((u = (s = t == null ? void 0 : t.locale) == null ? void 0 : s.options) == null ? void 0 : u.weekStartsOn) ?? n.weekStartsOn ?? ((l = (c = n.locale) == null ? void 0 : c.options) == null ? void 0 : l.weekStartsOn) ?? 0, a = A(e), i = a.getDay(), o = (i < r ? 7 : 0) + i - r;
  return a.setDate(a.getDate() - o), a.setHours(0, 0, 0, 0), a;
}
const ir = {
  lessThanXSeconds: {
    one: "less than a second",
    other: "less than {{count}} seconds"
  },
  xSeconds: {
    one: "1 second",
    other: "{{count}} seconds"
  },
  halfAMinute: "half a minute",
  lessThanXMinutes: {
    one: "less than a minute",
    other: "less than {{count}} minutes"
  },
  xMinutes: {
    one: "1 minute",
    other: "{{count}} minutes"
  },
  aboutXHours: {
    one: "about 1 hour",
    other: "about {{count}} hours"
  },
  xHours: {
    one: "1 hour",
    other: "{{count}} hours"
  },
  xDays: {
    one: "1 day",
    other: "{{count}} days"
  },
  aboutXWeeks: {
    one: "about 1 week",
    other: "about {{count}} weeks"
  },
  xWeeks: {
    one: "1 week",
    other: "{{count}} weeks"
  },
  aboutXMonths: {
    one: "about 1 month",
    other: "about {{count}} months"
  },
  xMonths: {
    one: "1 month",
    other: "{{count}} months"
  },
  aboutXYears: {
    one: "about 1 year",
    other: "about {{count}} years"
  },
  xYears: {
    one: "1 year",
    other: "{{count}} years"
  },
  overXYears: {
    one: "over 1 year",
    other: "over {{count}} years"
  },
  almostXYears: {
    one: "almost 1 year",
    other: "almost {{count}} years"
  }
}, or = (e, t, n) => {
  let r;
  const a = ir[e];
  return typeof a == "string" ? r = a : t === 1 ? r = a.one : r = a.other.replace("{{count}}", t.toString()), n != null && n.addSuffix ? n.comparison && n.comparison > 0 ? "in " + r : r + " ago" : r;
}, sr = {
  lastWeek: "'last' eeee 'at' p",
  yesterday: "'yesterday at' p",
  today: "'today at' p",
  tomorrow: "'tomorrow at' p",
  nextWeek: "eeee 'at' p",
  other: "P"
}, ur = (e, t, n, r) => sr[e], cr = {
  narrow: ["B", "A"],
  abbreviated: ["BC", "AD"],
  wide: ["Before Christ", "Anno Domini"]
}, lr = {
  narrow: ["1", "2", "3", "4"],
  abbreviated: ["Q1", "Q2", "Q3", "Q4"],
  wide: ["1st quarter", "2nd quarter", "3rd quarter", "4th quarter"]
}, fr = {
  narrow: ["J", "F", "M", "A", "M", "J", "J", "A", "S", "O", "N", "D"],
  abbreviated: [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec"
  ],
  wide: [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December"
  ]
}, dr = {
  narrow: ["S", "M", "T", "W", "T", "F", "S"],
  short: ["Su", "Mo", "Tu", "We", "Th", "Fr", "Sa"],
  abbreviated: ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"],
  wide: [
    "Sunday",
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday"
  ]
}, hr = {
  narrow: {
    am: "a",
    pm: "p",
    midnight: "mi",
    noon: "n",
    morning: "morning",
    afternoon: "afternoon",
    evening: "evening",
    night: "night"
  },
  abbreviated: {
    am: "AM",
    pm: "PM",
    midnight: "midnight",
    noon: "noon",
    morning: "morning",
    afternoon: "afternoon",
    evening: "evening",
    night: "night"
  },
  wide: {
    am: "a.m.",
    pm: "p.m.",
    midnight: "midnight",
    noon: "noon",
    morning: "morning",
    afternoon: "afternoon",
    evening: "evening",
    night: "night"
  }
}, gr = {
  narrow: {
    am: "a",
    pm: "p",
    midnight: "mi",
    noon: "n",
    morning: "in the morning",
    afternoon: "in the afternoon",
    evening: "in the evening",
    night: "at night"
  },
  abbreviated: {
    am: "AM",
    pm: "PM",
    midnight: "midnight",
    noon: "noon",
    morning: "in the morning",
    afternoon: "in the afternoon",
    evening: "in the evening",
    night: "at night"
  },
  wide: {
    am: "a.m.",
    pm: "p.m.",
    midnight: "midnight",
    noon: "noon",
    morning: "in the morning",
    afternoon: "in the afternoon",
    evening: "in the evening",
    night: "at night"
  }
}, mr = (e, t) => {
  const n = Number(e), r = n % 100;
  if (r > 20 || r < 10)
    switch (r % 10) {
      case 1:
        return n + "st";
      case 2:
        return n + "nd";
      case 3:
        return n + "rd";
    }
  return n + "th";
}, yr = {
  ordinalNumber: mr,
  era: re({
    values: cr,
    defaultWidth: "wide"
  }),
  quarter: re({
    values: lr,
    defaultWidth: "wide",
    argumentCallback: (e) => e - 1
  }),
  month: re({
    values: fr,
    defaultWidth: "wide"
  }),
  day: re({
    values: dr,
    defaultWidth: "wide"
  }),
  dayPeriod: re({
    values: hr,
    defaultWidth: "wide",
    formattingValues: gr,
    defaultFormattingWidth: "wide"
  })
}, pr = /^(\d+)(th|st|nd|rd)?/i, br = /\d+/i, wr = {
  narrow: /^(b|a)/i,
  abbreviated: /^(b\.?\s?c\.?|b\.?\s?c\.?\s?e\.?|a\.?\s?d\.?|c\.?\s?e\.?)/i,
  wide: /^(before christ|before common era|anno domini|common era)/i
}, Tr = {
  any: [/^b/i, /^(a|c)/i]
}, vr = {
  narrow: /^[1234]/i,
  abbreviated: /^q[1234]/i,
  wide: /^[1234](th|st|nd|rd)? quarter/i
}, Or = {
  any: [/1/i, /2/i, /3/i, /4/i]
}, Sr = {
  narrow: /^[jfmasond]/i,
  abbreviated: /^(jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)/i,
  wide: /^(january|february|march|april|may|june|july|august|september|october|november|december)/i
}, Mr = {
  narrow: [
    /^j/i,
    /^f/i,
    /^m/i,
    /^a/i,
    /^m/i,
    /^j/i,
    /^j/i,
    /^a/i,
    /^s/i,
    /^o/i,
    /^n/i,
    /^d/i
  ],
  any: [
    /^ja/i,
    /^f/i,
    /^mar/i,
    /^ap/i,
    /^may/i,
    /^jun/i,
    /^jul/i,
    /^au/i,
    /^s/i,
    /^o/i,
    /^n/i,
    /^d/i
  ]
}, _r = {
  narrow: /^[smtwf]/i,
  short: /^(su|mo|tu|we|th|fr|sa)/i,
  abbreviated: /^(sun|mon|tue|wed|thu|fri|sat)/i,
  wide: /^(sunday|monday|tuesday|wednesday|thursday|friday|saturday)/i
}, Pr = {
  narrow: [/^s/i, /^m/i, /^t/i, /^w/i, /^t/i, /^f/i, /^s/i],
  any: [/^su/i, /^m/i, /^tu/i, /^w/i, /^th/i, /^f/i, /^sa/i]
}, xr = {
  narrow: /^(a|p|mi|n|(in the|at) (morning|afternoon|evening|night))/i,
  any: /^([ap]\.?\s?m\.?|midnight|noon|(in the|at) (morning|afternoon|evening|night))/i
}, Ar = {
  any: {
    am: /^a/i,
    pm: /^p/i,
    midnight: /^mi/i,
    noon: /^no/i,
    morning: /morning/i,
    afternoon: /afternoon/i,
    evening: /evening/i,
    night: /night/i
  }
}, Er = {
  ordinalNumber: rr({
    matchPattern: pr,
    parsePattern: br,
    valueCallback: (e) => parseInt(e, 10)
  }),
  era: ae({
    matchPatterns: wr,
    defaultMatchWidth: "wide",
    parsePatterns: Tr,
    defaultParseWidth: "any"
  }),
  quarter: ae({
    matchPatterns: vr,
    defaultMatchWidth: "wide",
    parsePatterns: Or,
    defaultParseWidth: "any",
    valueCallback: (e) => e + 1
  }),
  month: ae({
    matchPatterns: Sr,
    defaultMatchWidth: "wide",
    parsePatterns: Mr,
    defaultParseWidth: "any"
  }),
  day: ae({
    matchPatterns: _r,
    defaultMatchWidth: "wide",
    parsePatterns: Pr,
    defaultParseWidth: "any"
  }),
  dayPeriod: ae({
    matchPatterns: xr,
    defaultMatchWidth: "any",
    parsePatterns: Ar,
    defaultParseWidth: "any"
  })
}, $r = {
  full: "EEEE, MMMM do, y",
  long: "MMMM do, y",
  medium: "MMM d, y",
  short: "MM/dd/yyyy"
}, Cr = {
  full: "h:mm:ss a zzzz",
  long: "h:mm:ss a z",
  medium: "h:mm:ss a",
  short: "h:mm a"
}, Dr = {
  full: "{{date}} 'at' {{time}}",
  long: "{{date}} 'at' {{time}}",
  medium: "{{date}}, {{time}}",
  short: "{{date}}, {{time}}"
}, Ir = {
  date: Ie({
    formats: $r,
    defaultWidth: "full"
  }),
  time: Ie({
    formats: Cr,
    defaultWidth: "full"
  }),
  dateTime: Ie({
    formats: Dr,
    defaultWidth: "full"
  })
}, Qt = {
  code: "en-US",
  formatDistance: or,
  formatLong: Ir,
  formatRelative: ur,
  localize: yr,
  match: Er,
  options: {
    weekStartsOn: 0,
    firstWeekContainsDate: 1
  }
};
var Xt = typeof global == "object" && global && global.Object === Object && global, Fr = typeof self == "object" && self && self.Object === Object && self, D = Xt || Fr || Function("return this")(), P = D.Symbol, kt = Object.prototype, Lr = kt.hasOwnProperty, Rr = kt.toString, ie = P ? P.toStringTag : void 0;
function jr(e) {
  var t = Lr.call(e, ie), n = e[ie];
  try {
    e[ie] = void 0;
    var r = !0;
  } catch {
  }
  var a = Rr.call(e);
  return r && (t ? e[ie] = n : delete e[ie]), a;
}
var Nr = Object.prototype, Wr = Nr.toString;
function Yr(e) {
  return Wr.call(e);
}
var Gr = "[object Null]", Br = "[object Undefined]", dt = P ? P.toStringTag : void 0;
function U(e) {
  return e == null ? e === void 0 ? Br : Gr : dt && dt in Object(e) ? jr(e) : Yr(e);
}
function C(e) {
  return e != null && typeof e == "object";
}
var Hr = "[object Symbol]";
function Qe(e) {
  return typeof e == "symbol" || C(e) && U(e) == Hr;
}
function qr(e, t) {
  for (var n = -1, r = e == null ? 0 : e.length, a = Array(r); ++n < r; )
    a[n] = t(e[n], n, e);
  return a;
}
var _ = Array.isArray, Ur = 1 / 0, ht = P ? P.prototype : void 0, gt = ht ? ht.toString : void 0;
function Vt(e) {
  if (typeof e == "string")
    return e;
  if (_(e))
    return qr(e, Vt) + "";
  if (Qe(e))
    return gt ? gt.call(e) : "";
  var t = e + "";
  return t == "0" && 1 / e == -Ur ? "-0" : t;
}
function I(e) {
  var t = typeof e;
  return e != null && (t == "object" || t == "function");
}
function Xe(e) {
  return e;
}
var zr = "[object AsyncFunction]", Qr = "[object Function]", Xr = "[object GeneratorFunction]", kr = "[object Proxy]";
function ke(e) {
  if (!I(e))
    return !1;
  var t = U(e);
  return t == Qr || t == Xr || t == zr || t == kr;
}
var Fe = D["__core-js_shared__"], mt = function() {
  var e = /[^.]+$/.exec(Fe && Fe.keys && Fe.keys.IE_PROTO || "");
  return e ? "Symbol(src)_1." + e : "";
}();
function Vr(e) {
  return !!mt && mt in e;
}
var Zr = Function.prototype, Kr = Zr.toString;
function z(e) {
  if (e != null) {
    try {
      return Kr.call(e);
    } catch {
    }
    try {
      return e + "";
    } catch {
    }
  }
  return "";
}
var Jr = /[\\^$.*+?()[\]{}|]/g, ea = /^\[object .+?Constructor\]$/, ta = Function.prototype, na = Object.prototype, ra = ta.toString, aa = na.hasOwnProperty, ia = RegExp(
  "^" + ra.call(aa).replace(Jr, "\\$&").replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, "$1.*?") + "$"
);
function oa(e) {
  if (!I(e) || Vr(e))
    return !1;
  var t = ke(e) ? ia : ea;
  return t.test(z(e));
}
function sa(e, t) {
  return e == null ? void 0 : e[t];
}
function Q(e, t) {
  var n = sa(e, t);
  return oa(n) ? n : void 0;
}
var We = Q(D, "WeakMap"), yt = Object.create, ua = /* @__PURE__ */ function() {
  function e() {
  }
  return function(t) {
    if (!I(t))
      return {};
    if (yt)
      return yt(t);
    e.prototype = t;
    var n = new e();
    return e.prototype = void 0, n;
  };
}();
function ca(e, t, n) {
  switch (n.length) {
    case 0:
      return e.call(t);
    case 1:
      return e.call(t, n[0]);
    case 2:
      return e.call(t, n[0], n[1]);
    case 3:
      return e.call(t, n[0], n[1], n[2]);
  }
  return e.apply(t, n);
}
function la() {
}
function Zt(e, t) {
  var n = -1, r = e.length;
  for (t || (t = Array(r)); ++n < r; )
    t[n] = e[n];
  return t;
}
var fa = 800, da = 16, ha = Date.now;
function ga(e) {
  var t = 0, n = 0;
  return function() {
    var r = ha(), a = da - (r - n);
    if (n = r, a > 0) {
      if (++t >= fa)
        return arguments[0];
    } else
      t = 0;
    return e.apply(void 0, arguments);
  };
}
function ma(e) {
  return function() {
    return e;
  };
}
var Oe = function() {
  try {
    var e = Q(Object, "defineProperty");
    return e({}, "", {}), e;
  } catch {
  }
}(), ya = Oe ? function(e, t) {
  return Oe(e, "toString", {
    configurable: !0,
    enumerable: !1,
    value: ma(t),
    writable: !0
  });
} : Xe;
const pa = ya;
var ba = ga(pa);
function wa(e, t) {
  for (var n = -1, r = e == null ? 0 : e.length; ++n < r && t(e[n], n, e) !== !1; )
    ;
  return e;
}
function Ta(e, t, n, r) {
  for (var a = e.length, i = n + (r ? 1 : -1); r ? i-- : ++i < a; )
    if (t(e[i], i, e))
      return i;
  return -1;
}
function va(e) {
  return e !== e;
}
function Oa(e, t, n) {
  for (var r = n - 1, a = e.length; ++r < a; )
    if (e[r] === t)
      return r;
  return -1;
}
function Sa(e, t, n) {
  return t === t ? Oa(e, t, n) : Ta(e, va, n);
}
function Ma(e, t) {
  var n = e == null ? 0 : e.length;
  return !!n && Sa(e, t, 0) > -1;
}
var _a = 9007199254740991, Pa = /^(?:0|[1-9]\d*)$/;
function Ve(e, t) {
  var n = typeof e;
  return t = t ?? _a, !!t && (n == "number" || n != "symbol" && Pa.test(e)) && e > -1 && e % 1 == 0 && e < t;
}
function Ze(e, t, n) {
  t == "__proto__" && Oe ? Oe(e, t, {
    configurable: !0,
    enumerable: !0,
    value: n,
    writable: !0
  }) : e[t] = n;
}
function de(e, t) {
  return e === t || e !== e && t !== t;
}
var xa = Object.prototype, Aa = xa.hasOwnProperty;
function Kt(e, t, n) {
  var r = e[t];
  (!(Aa.call(e, t) && de(r, n)) || n === void 0 && !(t in e)) && Ze(e, t, n);
}
function he(e, t, n, r) {
  var a = !n;
  n || (n = {});
  for (var i = -1, o = t.length; ++i < o; ) {
    var s = t[i], u = r ? r(n[s], e[s], s, n, e) : void 0;
    u === void 0 && (u = e[s]), a ? Ze(n, s, u) : Kt(n, s, u);
  }
  return n;
}
var pt = Math.max;
function Ea(e, t, n) {
  return t = pt(t === void 0 ? e.length - 1 : t, 0), function() {
    for (var r = arguments, a = -1, i = pt(r.length - t, 0), o = Array(i); ++a < i; )
      o[a] = r[t + a];
    a = -1;
    for (var s = Array(t + 1); ++a < t; )
      s[a] = r[a];
    return s[t] = n(o), ca(e, this, s);
  };
}
function Jt(e, t) {
  return ba(Ea(e, t, Xe), e + "");
}
var $a = 9007199254740991;
function Ke(e) {
  return typeof e == "number" && e > -1 && e % 1 == 0 && e <= $a;
}
function ge(e) {
  return e != null && Ke(e.length) && !ke(e);
}
function Ca(e, t, n) {
  if (!I(n))
    return !1;
  var r = typeof t;
  return (r == "number" ? ge(n) && Ve(t, n.length) : r == "string" && t in n) ? de(n[t], e) : !1;
}
function Da(e) {
  return Jt(function(t, n) {
    var r = -1, a = n.length, i = a > 1 ? n[a - 1] : void 0, o = a > 2 ? n[2] : void 0;
    for (i = e.length > 3 && typeof i == "function" ? (a--, i) : void 0, o && Ca(n[0], n[1], o) && (i = a < 3 ? void 0 : i, a = 1), t = Object(t); ++r < a; ) {
      var s = n[r];
      s && e(t, s, r, i);
    }
    return t;
  });
}
var Ia = Object.prototype;
function xe(e) {
  var t = e && e.constructor, n = typeof t == "function" && t.prototype || Ia;
  return e === n;
}
function Fa(e, t) {
  for (var n = -1, r = Array(e); ++n < e; )
    r[n] = t(n);
  return r;
}
var La = "[object Arguments]";
function bt(e) {
  return C(e) && U(e) == La;
}
var en = Object.prototype, Ra = en.hasOwnProperty, ja = en.propertyIsEnumerable, Na = bt(/* @__PURE__ */ function() {
  return arguments;
}()) ? bt : function(e) {
  return C(e) && Ra.call(e, "callee") && !ja.call(e, "callee");
};
const Z = Na;
function Wa() {
  return !1;
}
var tn = typeof exports == "object" && exports && !exports.nodeType && exports, wt = tn && typeof module == "object" && module && !module.nodeType && module, Ya = wt && wt.exports === tn, Tt = Ya ? D.Buffer : void 0, Ga = Tt ? Tt.isBuffer : void 0, Ba = Ga || Wa;
const K = Ba;
var Ha = "[object Arguments]", qa = "[object Array]", Ua = "[object Boolean]", za = "[object Date]", Qa = "[object Error]", Xa = "[object Function]", ka = "[object Map]", Va = "[object Number]", Za = "[object Object]", Ka = "[object RegExp]", Ja = "[object Set]", ei = "[object String]", ti = "[object WeakMap]", ni = "[object ArrayBuffer]", ri = "[object DataView]", ai = "[object Float32Array]", ii = "[object Float64Array]", oi = "[object Int8Array]", si = "[object Int16Array]", ui = "[object Int32Array]", ci = "[object Uint8Array]", li = "[object Uint8ClampedArray]", fi = "[object Uint16Array]", di = "[object Uint32Array]", p = {};
p[ai] = p[ii] = p[oi] = p[si] = p[ui] = p[ci] = p[li] = p[fi] = p[di] = !0;
p[Ha] = p[qa] = p[ni] = p[Ua] = p[ri] = p[za] = p[Qa] = p[Xa] = p[ka] = p[Va] = p[Za] = p[Ka] = p[Ja] = p[ei] = p[ti] = !1;
function hi(e) {
  return C(e) && Ke(e.length) && !!p[U(e)];
}
function Je(e) {
  return function(t) {
    return e(t);
  };
}
var nn = typeof exports == "object" && exports && !exports.nodeType && exports, se = nn && typeof module == "object" && module && !module.nodeType && module, gi = se && se.exports === nn, Le = gi && Xt.process, J = function() {
  try {
    var e = se && se.require && se.require("util").types;
    return e || Le && Le.binding && Le.binding("util");
  } catch {
  }
}(), vt = J && J.isTypedArray, mi = vt ? Je(vt) : hi;
const Ae = mi;
var yi = Object.prototype, pi = yi.hasOwnProperty;
function rn(e, t) {
  var n = _(e), r = !n && Z(e), a = !n && !r && K(e), i = !n && !r && !a && Ae(e), o = n || r || a || i, s = o ? Fa(e.length, String) : [], u = s.length;
  for (var c in e)
    (t || pi.call(e, c)) && !(o && // Safari 9 has enumerable `arguments.length` in strict mode.
    (c == "length" || // Node.js 0.10 has enumerable non-index properties on buffers.
    a && (c == "offset" || c == "parent") || // PhantomJS 2 has enumerable non-index properties on typed arrays.
    i && (c == "buffer" || c == "byteLength" || c == "byteOffset") || // Skip index properties.
    Ve(c, u))) && s.push(c);
  return s;
}
function an(e, t) {
  return function(n) {
    return e(t(n));
  };
}
var bi = an(Object.keys, Object), wi = Object.prototype, Ti = wi.hasOwnProperty;
function on(e) {
  if (!xe(e))
    return bi(e);
  var t = [];
  for (var n in Object(e))
    Ti.call(e, n) && n != "constructor" && t.push(n);
  return t;
}
function Ee(e) {
  return ge(e) ? rn(e) : on(e);
}
function vi(e) {
  var t = [];
  if (e != null)
    for (var n in Object(e))
      t.push(n);
  return t;
}
var Oi = Object.prototype, Si = Oi.hasOwnProperty;
function Mi(e) {
  if (!I(e))
    return vi(e);
  var t = xe(e), n = [];
  for (var r in e)
    r == "constructor" && (t || !Si.call(e, r)) || n.push(r);
  return n;
}
function me(e) {
  return ge(e) ? rn(e, !0) : Mi(e);
}
var _i = /\.|\[(?:[^[\]]*|(["'])(?:(?!\1)[^\\]|\\.)*?\1)\]/, Pi = /^\w*$/;
function et(e, t) {
  if (_(e))
    return !1;
  var n = typeof e;
  return n == "number" || n == "symbol" || n == "boolean" || e == null || Qe(e) ? !0 : Pi.test(e) || !_i.test(e) || t != null && e in Object(t);
}
var ce = Q(Object, "create");
function xi() {
  this.__data__ = ce ? ce(null) : {}, this.size = 0;
}
function Ai(e) {
  var t = this.has(e) && delete this.__data__[e];
  return this.size -= t ? 1 : 0, t;
}
var Ei = "__lodash_hash_undefined__", $i = Object.prototype, Ci = $i.hasOwnProperty;
function Di(e) {
  var t = this.__data__;
  if (ce) {
    var n = t[e];
    return n === Ei ? void 0 : n;
  }
  return Ci.call(t, e) ? t[e] : void 0;
}
var Ii = Object.prototype, Fi = Ii.hasOwnProperty;
function Li(e) {
  var t = this.__data__;
  return ce ? t[e] !== void 0 : Fi.call(t, e);
}
var Ri = "__lodash_hash_undefined__";
function ji(e, t) {
  var n = this.__data__;
  return this.size += this.has(e) ? 0 : 1, n[e] = ce && t === void 0 ? Ri : t, this;
}
function B(e) {
  var t = -1, n = e == null ? 0 : e.length;
  for (this.clear(); ++t < n; ) {
    var r = e[t];
    this.set(r[0], r[1]);
  }
}
B.prototype.clear = xi;
B.prototype.delete = Ai;
B.prototype.get = Di;
B.prototype.has = Li;
B.prototype.set = ji;
function Ni() {
  this.__data__ = [], this.size = 0;
}
function $e(e, t) {
  for (var n = e.length; n--; )
    if (de(e[n][0], t))
      return n;
  return -1;
}
var Wi = Array.prototype, Yi = Wi.splice;
function Gi(e) {
  var t = this.__data__, n = $e(t, e);
  if (n < 0)
    return !1;
  var r = t.length - 1;
  return n == r ? t.pop() : Yi.call(t, n, 1), --this.size, !0;
}
function Bi(e) {
  var t = this.__data__, n = $e(t, e);
  return n < 0 ? void 0 : t[n][1];
}
function Hi(e) {
  return $e(this.__data__, e) > -1;
}
function qi(e, t) {
  var n = this.__data__, r = $e(n, e);
  return r < 0 ? (++this.size, n.push([e, t])) : n[r][1] = t, this;
}
function F(e) {
  var t = -1, n = e == null ? 0 : e.length;
  for (this.clear(); ++t < n; ) {
    var r = e[t];
    this.set(r[0], r[1]);
  }
}
F.prototype.clear = Ni;
F.prototype.delete = Gi;
F.prototype.get = Bi;
F.prototype.has = Hi;
F.prototype.set = qi;
var le = Q(D, "Map");
function Ui() {
  this.size = 0, this.__data__ = {
    hash: new B(),
    map: new (le || F)(),
    string: new B()
  };
}
function zi(e) {
  var t = typeof e;
  return t == "string" || t == "number" || t == "symbol" || t == "boolean" ? e !== "__proto__" : e === null;
}
function Ce(e, t) {
  var n = e.__data__;
  return zi(t) ? n[typeof t == "string" ? "string" : "hash"] : n.map;
}
function Qi(e) {
  var t = Ce(this, e).delete(e);
  return this.size -= t ? 1 : 0, t;
}
function Xi(e) {
  return Ce(this, e).get(e);
}
function ki(e) {
  return Ce(this, e).has(e);
}
function Vi(e, t) {
  var n = Ce(this, e), r = n.size;
  return n.set(e, t), this.size += n.size == r ? 0 : 1, this;
}
function L(e) {
  var t = -1, n = e == null ? 0 : e.length;
  for (this.clear(); ++t < n; ) {
    var r = e[t];
    this.set(r[0], r[1]);
  }
}
L.prototype.clear = Ui;
L.prototype.delete = Qi;
L.prototype.get = Xi;
L.prototype.has = ki;
L.prototype.set = Vi;
var Zi = "Expected a function";
function tt(e, t) {
  if (typeof e != "function" || t != null && typeof t != "function")
    throw new TypeError(Zi);
  var n = function() {
    var r = arguments, a = t ? t.apply(this, r) : r[0], i = n.cache;
    if (i.has(a))
      return i.get(a);
    var o = e.apply(this, r);
    return n.cache = i.set(a, o) || i, o;
  };
  return n.cache = new (tt.Cache || L)(), n;
}
tt.Cache = L;
var Ki = 500;
function Ji(e) {
  var t = tt(e, function(r) {
    return n.size === Ki && n.clear(), r;
  }), n = t.cache;
  return t;
}
var eo = /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|$))/g, to = /\\(\\)?/g, no = Ji(function(e) {
  var t = [];
  return e.charCodeAt(0) === 46 && t.push(""), e.replace(eo, function(n, r, a, i) {
    t.push(a ? i.replace(to, "$1") : r || n);
  }), t;
});
function ro(e) {
  return e == null ? "" : Vt(e);
}
function sn(e, t) {
  return _(e) ? e : et(e, t) ? [e] : no(ro(e));
}
var ao = 1 / 0;
function De(e) {
  if (typeof e == "string" || Qe(e))
    return e;
  var t = e + "";
  return t == "0" && 1 / e == -ao ? "-0" : t;
}
function un(e, t) {
  t = sn(t, e);
  for (var n = 0, r = t.length; e != null && n < r; )
    e = e[De(t[n++])];
  return n && n == r ? e : void 0;
}
function io(e, t, n) {
  var r = e == null ? void 0 : un(e, t);
  return r === void 0 ? n : r;
}
function nt(e, t) {
  for (var n = -1, r = t.length, a = e.length; ++n < r; )
    e[a + n] = t[n];
  return e;
}
var Ot = P ? P.isConcatSpreadable : void 0;
function oo(e) {
  return _(e) || Z(e) || !!(Ot && e && e[Ot]);
}
function cn(e, t, n, r, a) {
  var i = -1, o = e.length;
  for (n || (n = oo), a || (a = []); ++i < o; ) {
    var s = e[i];
    t > 0 && n(s) ? t > 1 ? cn(s, t - 1, n, r, a) : nt(a, s) : r || (a[a.length] = s);
  }
  return a;
}
var so = an(Object.getPrototypeOf, Object);
const rt = so;
var uo = "[object Object]", co = Function.prototype, lo = Object.prototype, ln = co.toString, fo = lo.hasOwnProperty, ho = ln.call(Object);
function go(e) {
  if (!C(e) || U(e) != uo)
    return !1;
  var t = rt(e);
  if (t === null)
    return !0;
  var n = fo.call(t, "constructor") && t.constructor;
  return typeof n == "function" && n instanceof n && ln.call(n) == ho;
}
function mo() {
  this.__data__ = new F(), this.size = 0;
}
function yo(e) {
  var t = this.__data__, n = t.delete(e);
  return this.size = t.size, n;
}
function po(e) {
  return this.__data__.get(e);
}
function bo(e) {
  return this.__data__.has(e);
}
var wo = 200;
function To(e, t) {
  var n = this.__data__;
  if (n instanceof F) {
    var r = n.__data__;
    if (!le || r.length < wo - 1)
      return r.push([e, t]), this.size = ++n.size, this;
    n = this.__data__ = new L(r);
  }
  return n.set(e, t), this.size = n.size, this;
}
function x(e) {
  var t = this.__data__ = new F(e);
  this.size = t.size;
}
x.prototype.clear = mo;
x.prototype.delete = yo;
x.prototype.get = po;
x.prototype.has = bo;
x.prototype.set = To;
function vo(e, t) {
  return e && he(t, Ee(t), e);
}
function Oo(e, t) {
  return e && he(t, me(t), e);
}
var fn = typeof exports == "object" && exports && !exports.nodeType && exports, St = fn && typeof module == "object" && module && !module.nodeType && module, So = St && St.exports === fn, Mt = So ? D.Buffer : void 0, _t = Mt ? Mt.allocUnsafe : void 0;
function dn(e, t) {
  if (t)
    return e.slice();
  var n = e.length, r = _t ? _t(n) : new e.constructor(n);
  return e.copy(r), r;
}
function Mo(e, t) {
  for (var n = -1, r = e == null ? 0 : e.length, a = 0, i = []; ++n < r; ) {
    var o = e[n];
    t(o, n, e) && (i[a++] = o);
  }
  return i;
}
function hn() {
  return [];
}
var _o = Object.prototype, Po = _o.propertyIsEnumerable, Pt = Object.getOwnPropertySymbols, at = Pt ? function(e) {
  return e == null ? [] : (e = Object(e), Mo(Pt(e), function(t) {
    return Po.call(e, t);
  }));
} : hn;
function xo(e, t) {
  return he(e, at(e), t);
}
var Ao = Object.getOwnPropertySymbols, gn = Ao ? function(e) {
  for (var t = []; e; )
    nt(t, at(e)), e = rt(e);
  return t;
} : hn;
function Eo(e, t) {
  return he(e, gn(e), t);
}
function mn(e, t, n) {
  var r = t(e);
  return _(e) ? r : nt(r, n(e));
}
function Ye(e) {
  return mn(e, Ee, at);
}
function $o(e) {
  return mn(e, me, gn);
}
var Ge = Q(D, "DataView"), Be = Q(D, "Promise"), V = Q(D, "Set"), xt = "[object Map]", Co = "[object Object]", At = "[object Promise]", Et = "[object Set]", $t = "[object WeakMap]", Ct = "[object DataView]", Do = z(Ge), Io = z(le), Fo = z(Be), Lo = z(V), Ro = z(We), W = U;
(Ge && W(new Ge(new ArrayBuffer(1))) != Ct || le && W(new le()) != xt || Be && W(Be.resolve()) != At || V && W(new V()) != Et || We && W(new We()) != $t) && (W = function(e) {
  var t = U(e), n = t == Co ? e.constructor : void 0, r = n ? z(n) : "";
  if (r)
    switch (r) {
      case Do:
        return Ct;
      case Io:
        return xt;
      case Fo:
        return At;
      case Lo:
        return Et;
      case Ro:
        return $t;
    }
  return t;
});
const ee = W;
var jo = Object.prototype, No = jo.hasOwnProperty;
function Wo(e) {
  var t = e.length, n = new e.constructor(t);
  return t && typeof e[0] == "string" && No.call(e, "index") && (n.index = e.index, n.input = e.input), n;
}
var Se = D.Uint8Array;
function it(e) {
  var t = new e.constructor(e.byteLength);
  return new Se(t).set(new Se(e)), t;
}
function Yo(e, t) {
  var n = t ? it(e.buffer) : e.buffer;
  return new e.constructor(n, e.byteOffset, e.byteLength);
}
var Go = /\w*$/;
function Bo(e) {
  var t = new e.constructor(e.source, Go.exec(e));
  return t.lastIndex = e.lastIndex, t;
}
var Dt = P ? P.prototype : void 0, It = Dt ? Dt.valueOf : void 0;
function Ho(e) {
  return It ? Object(It.call(e)) : {};
}
function yn(e, t) {
  var n = t ? it(e.buffer) : e.buffer;
  return new e.constructor(n, e.byteOffset, e.length);
}
var qo = "[object Boolean]", Uo = "[object Date]", zo = "[object Map]", Qo = "[object Number]", Xo = "[object RegExp]", ko = "[object Set]", Vo = "[object String]", Zo = "[object Symbol]", Ko = "[object ArrayBuffer]", Jo = "[object DataView]", es = "[object Float32Array]", ts = "[object Float64Array]", ns = "[object Int8Array]", rs = "[object Int16Array]", as = "[object Int32Array]", is = "[object Uint8Array]", os = "[object Uint8ClampedArray]", ss = "[object Uint16Array]", us = "[object Uint32Array]";
function cs(e, t, n) {
  var r = e.constructor;
  switch (t) {
    case Ko:
      return it(e);
    case qo:
    case Uo:
      return new r(+e);
    case Jo:
      return Yo(e, n);
    case es:
    case ts:
    case ns:
    case rs:
    case as:
    case is:
    case os:
    case ss:
    case us:
      return yn(e, n);
    case zo:
      return new r();
    case Qo:
    case Vo:
      return new r(e);
    case Xo:
      return Bo(e);
    case ko:
      return new r();
    case Zo:
      return Ho(e);
  }
}
function pn(e) {
  return typeof e.constructor == "function" && !xe(e) ? ua(rt(e)) : {};
}
var ls = "[object Map]";
function fs(e) {
  return C(e) && ee(e) == ls;
}
var Ft = J && J.isMap, ds = Ft ? Je(Ft) : fs, hs = "[object Set]";
function gs(e) {
  return C(e) && ee(e) == hs;
}
var Lt = J && J.isSet, ms = Lt ? Je(Lt) : gs, ys = 1, ps = 2, bs = 4, bn = "[object Arguments]", ws = "[object Array]", Ts = "[object Boolean]", vs = "[object Date]", Os = "[object Error]", wn = "[object Function]", Ss = "[object GeneratorFunction]", Ms = "[object Map]", _s = "[object Number]", Tn = "[object Object]", Ps = "[object RegExp]", xs = "[object Set]", As = "[object String]", Es = "[object Symbol]", $s = "[object WeakMap]", Cs = "[object ArrayBuffer]", Ds = "[object DataView]", Is = "[object Float32Array]", Fs = "[object Float64Array]", Ls = "[object Int8Array]", Rs = "[object Int16Array]", js = "[object Int32Array]", Ns = "[object Uint8Array]", Ws = "[object Uint8ClampedArray]", Ys = "[object Uint16Array]", Gs = "[object Uint32Array]", y = {};
y[bn] = y[ws] = y[Cs] = y[Ds] = y[Ts] = y[vs] = y[Is] = y[Fs] = y[Ls] = y[Rs] = y[js] = y[Ms] = y[_s] = y[Tn] = y[Ps] = y[xs] = y[As] = y[Es] = y[Ns] = y[Ws] = y[Ys] = y[Gs] = !0;
y[Os] = y[wn] = y[$s] = !1;
function ve(e, t, n, r, a, i) {
  var o, s = t & ys, u = t & ps, c = t & bs;
  if (n && (o = a ? n(e, r, a, i) : n(e)), o !== void 0)
    return o;
  if (!I(e))
    return e;
  var l = _(e);
  if (l) {
    if (o = Wo(e), !s)
      return Zt(e, o);
  } else {
    var f = ee(e), d = f == wn || f == Ss;
    if (K(e))
      return dn(e, s);
    if (f == Tn || f == bn || d && !a) {
      if (o = u || d ? {} : pn(e), !s)
        return u ? Eo(e, Oo(o, e)) : xo(e, vo(o, e));
    } else {
      if (!y[f])
        return a ? e : {};
      o = cs(e, f, s);
    }
  }
  i || (i = new x());
  var m = i.get(e);
  if (m)
    return m;
  i.set(e, o), ms(e) ? e.forEach(function(w) {
    o.add(ve(w, t, n, w, e, i));
  }) : ds(e) && e.forEach(function(w, O) {
    o.set(O, ve(w, t, n, O, e, i));
  });
  var T = c ? u ? $o : Ye : u ? me : Ee, v = l ? void 0 : T(e);
  return wa(v || e, function(w, O) {
    v && (O = w, w = e[O]), Kt(o, O, ve(w, t, n, O, e, i));
  }), o;
}
var Bs = 1, Hs = 4;
function qs(e) {
  return ve(e, Bs | Hs);
}
var Us = "__lodash_hash_undefined__";
function zs(e) {
  return this.__data__.set(e, Us), this;
}
function Qs(e) {
  return this.__data__.has(e);
}
function fe(e) {
  var t = -1, n = e == null ? 0 : e.length;
  for (this.__data__ = new L(); ++t < n; )
    this.add(e[t]);
}
fe.prototype.add = fe.prototype.push = zs;
fe.prototype.has = Qs;
function Xs(e, t) {
  for (var n = -1, r = e == null ? 0 : e.length; ++n < r; )
    if (t(e[n], n, e))
      return !0;
  return !1;
}
function vn(e, t) {
  return e.has(t);
}
var ks = 1, Vs = 2;
function On(e, t, n, r, a, i) {
  var o = n & ks, s = e.length, u = t.length;
  if (s != u && !(o && u > s))
    return !1;
  var c = i.get(e), l = i.get(t);
  if (c && l)
    return c == t && l == e;
  var f = -1, d = !0, m = n & Vs ? new fe() : void 0;
  for (i.set(e, t), i.set(t, e); ++f < s; ) {
    var T = e[f], v = t[f];
    if (r)
      var w = o ? r(v, T, f, t, e, i) : r(T, v, f, e, t, i);
    if (w !== void 0) {
      if (w)
        continue;
      d = !1;
      break;
    }
    if (m) {
      if (!Xs(t, function(O, b) {
        if (!vn(m, b) && (T === O || a(T, O, n, r, i)))
          return m.push(b);
      })) {
        d = !1;
        break;
      }
    } else if (!(T === v || a(T, v, n, r, i))) {
      d = !1;
      break;
    }
  }
  return i.delete(e), i.delete(t), d;
}
function Zs(e) {
  var t = -1, n = Array(e.size);
  return e.forEach(function(r, a) {
    n[++t] = [a, r];
  }), n;
}
function ot(e) {
  var t = -1, n = Array(e.size);
  return e.forEach(function(r) {
    n[++t] = r;
  }), n;
}
var Ks = 1, Js = 2, eu = "[object Boolean]", tu = "[object Date]", nu = "[object Error]", ru = "[object Map]", au = "[object Number]", iu = "[object RegExp]", ou = "[object Set]", su = "[object String]", uu = "[object Symbol]", cu = "[object ArrayBuffer]", lu = "[object DataView]", Rt = P ? P.prototype : void 0, Re = Rt ? Rt.valueOf : void 0;
function fu(e, t, n, r, a, i, o) {
  switch (n) {
    case lu:
      if (e.byteLength != t.byteLength || e.byteOffset != t.byteOffset)
        return !1;
      e = e.buffer, t = t.buffer;
    case cu:
      return !(e.byteLength != t.byteLength || !i(new Se(e), new Se(t)));
    case eu:
    case tu:
    case au:
      return de(+e, +t);
    case nu:
      return e.name == t.name && e.message == t.message;
    case iu:
    case su:
      return e == t + "";
    case ru:
      var s = Zs;
    case ou:
      var u = r & Ks;
      if (s || (s = ot), e.size != t.size && !u)
        return !1;
      var c = o.get(e);
      if (c)
        return c == t;
      r |= Js, o.set(e, t);
      var l = On(s(e), s(t), r, a, i, o);
      return o.delete(e), l;
    case uu:
      if (Re)
        return Re.call(e) == Re.call(t);
  }
  return !1;
}
var du = 1, hu = Object.prototype, gu = hu.hasOwnProperty;
function mu(e, t, n, r, a, i) {
  var o = n & du, s = Ye(e), u = s.length, c = Ye(t), l = c.length;
  if (u != l && !o)
    return !1;
  for (var f = u; f--; ) {
    var d = s[f];
    if (!(o ? d in t : gu.call(t, d)))
      return !1;
  }
  var m = i.get(e), T = i.get(t);
  if (m && T)
    return m == t && T == e;
  var v = !0;
  i.set(e, t), i.set(t, e);
  for (var w = o; ++f < u; ) {
    d = s[f];
    var O = e[d], b = t[d];
    if (r)
      var S = o ? r(b, O, d, t, e, i) : r(O, b, d, e, t, i);
    if (!(S === void 0 ? O === b || a(O, b, n, r, i) : S)) {
      v = !1;
      break;
    }
    w || (w = d == "constructor");
  }
  if (v && !w) {
    var E = e.constructor, R = t.constructor;
    E != R && "constructor" in e && "constructor" in t && !(typeof E == "function" && E instanceof E && typeof R == "function" && R instanceof R) && (v = !1);
  }
  return i.delete(e), i.delete(t), v;
}
var yu = 1, jt = "[object Arguments]", Nt = "[object Array]", be = "[object Object]", pu = Object.prototype, Wt = pu.hasOwnProperty;
function bu(e, t, n, r, a, i) {
  var o = _(e), s = _(t), u = o ? Nt : ee(e), c = s ? Nt : ee(t);
  u = u == jt ? be : u, c = c == jt ? be : c;
  var l = u == be, f = c == be, d = u == c;
  if (d && K(e)) {
    if (!K(t))
      return !1;
    o = !0, l = !1;
  }
  if (d && !l)
    return i || (i = new x()), o || Ae(e) ? On(e, t, n, r, a, i) : fu(e, t, u, n, r, a, i);
  if (!(n & yu)) {
    var m = l && Wt.call(e, "__wrapped__"), T = f && Wt.call(t, "__wrapped__");
    if (m || T) {
      var v = m ? e.value() : e, w = T ? t.value() : t;
      return i || (i = new x()), a(v, w, n, r, i);
    }
  }
  return d ? (i || (i = new x()), mu(e, t, n, r, a, i)) : !1;
}
function st(e, t, n, r, a) {
  return e === t ? !0 : e == null || t == null || !C(e) && !C(t) ? e !== e && t !== t : bu(e, t, n, r, st, a);
}
var wu = 1, Tu = 2;
function vu(e, t, n, r) {
  var a = n.length, i = a, o = !r;
  if (e == null)
    return !i;
  for (e = Object(e); a--; ) {
    var s = n[a];
    if (o && s[2] ? s[1] !== e[s[0]] : !(s[0] in e))
      return !1;
  }
  for (; ++a < i; ) {
    s = n[a];
    var u = s[0], c = e[u], l = s[1];
    if (o && s[2]) {
      if (c === void 0 && !(u in e))
        return !1;
    } else {
      var f = new x();
      if (r)
        var d = r(c, l, u, e, t, f);
      if (!(d === void 0 ? st(l, c, wu | Tu, r, f) : d))
        return !1;
    }
  }
  return !0;
}
function Sn(e) {
  return e === e && !I(e);
}
function Ou(e) {
  for (var t = Ee(e), n = t.length; n--; ) {
    var r = t[n], a = e[r];
    t[n] = [r, a, Sn(a)];
  }
  return t;
}
function Mn(e, t) {
  return function(n) {
    return n == null ? !1 : n[e] === t && (t !== void 0 || e in Object(n));
  };
}
function Su(e) {
  var t = Ou(e);
  return t.length == 1 && t[0][2] ? Mn(t[0][0], t[0][1]) : function(n) {
    return n === e || vu(n, e, t);
  };
}
function Mu(e, t) {
  return e != null && t in Object(e);
}
function _u(e, t, n) {
  t = sn(t, e);
  for (var r = -1, a = t.length, i = !1; ++r < a; ) {
    var o = De(t[r]);
    if (!(i = e != null && n(e, o)))
      break;
    e = e[o];
  }
  return i || ++r != a ? i : (a = e == null ? 0 : e.length, !!a && Ke(a) && Ve(o, a) && (_(e) || Z(e)));
}
function Pu(e, t) {
  return e != null && _u(e, t, Mu);
}
var xu = 1, Au = 2;
function Eu(e, t) {
  return et(e) && Sn(t) ? Mn(De(e), t) : function(n) {
    var r = io(n, e);
    return r === void 0 && r === t ? Pu(n, e) : st(t, r, xu | Au);
  };
}
function $u(e) {
  return function(t) {
    return t == null ? void 0 : t[e];
  };
}
function Cu(e) {
  return function(t) {
    return un(t, e);
  };
}
function Du(e) {
  return et(e) ? $u(De(e)) : Cu(e);
}
function Iu(e) {
  return typeof e == "function" ? e : e == null ? Xe : typeof e == "object" ? _(e) ? Eu(e[0], e[1]) : Su(e) : Du(e);
}
function Fu(e) {
  return function(t, n, r) {
    for (var a = -1, i = Object(t), o = r(t), s = o.length; s--; ) {
      var u = o[e ? s : ++a];
      if (n(i[u], u, i) === !1)
        break;
    }
    return t;
  };
}
var Lu = Fu();
function He(e, t, n) {
  (n !== void 0 && !de(e[t], n) || n === void 0 && !(t in e)) && Ze(e, t, n);
}
function qe(e) {
  return C(e) && ge(e);
}
function Ue(e, t) {
  if (!(t === "constructor" && typeof e[t] == "function") && t != "__proto__")
    return e[t];
}
function Ru(e) {
  return he(e, me(e));
}
function ju(e, t, n, r, a, i, o) {
  var s = Ue(e, n), u = Ue(t, n), c = o.get(u);
  if (c) {
    He(e, n, c);
    return;
  }
  var l = i ? i(s, u, n + "", e, t, o) : void 0, f = l === void 0;
  if (f) {
    var d = _(u), m = !d && K(u), T = !d && !m && Ae(u);
    l = u, d || m || T ? _(s) ? l = s : qe(s) ? l = Zt(s) : m ? (f = !1, l = dn(u, !0)) : T ? (f = !1, l = yn(u, !0)) : l = [] : go(u) || Z(u) ? (l = s, Z(s) ? l = Ru(s) : (!I(s) || ke(s)) && (l = pn(u))) : f = !1;
  }
  f && (o.set(u, l), a(l, u, r, i, o), o.delete(u)), He(e, n, l);
}
function _n(e, t, n, r, a) {
  e !== t && Lu(t, function(i, o) {
    if (a || (a = new x()), I(i))
      ju(e, t, o, n, _n, r, a);
    else {
      var s = r ? r(Ue(e, o), i, o + "", e, t, a) : void 0;
      s === void 0 && (s = i), He(e, o, s);
    }
  }, me);
}
function Nu(e, t, n) {
  for (var r = -1, a = e == null ? 0 : e.length; ++r < a; )
    if (n(t, e[r]))
      return !0;
  return !1;
}
function Wu(e) {
  var t = e == null ? 0 : e.length;
  return t ? e[t - 1] : void 0;
}
var Yu = "[object Map]", Gu = "[object Set]", Bu = Object.prototype, Hu = Bu.hasOwnProperty;
function qu(e) {
  if (e == null)
    return !0;
  if (ge(e) && (_(e) || typeof e == "string" || typeof e.splice == "function" || K(e) || Ae(e) || Z(e)))
    return !e.length;
  var t = ee(e);
  if (t == Yu || t == Gu)
    return !e.size;
  if (xe(e))
    return !on(e).length;
  for (var n in e)
    if (Hu.call(e, n))
      return !1;
  return !0;
}
var Uu = Da(function(e, t, n) {
  _n(e, t, n);
});
const h = Uu;
var zu = 1 / 0, Qu = V && 1 / ot(new V([, -0]))[1] == zu ? function(e) {
  return new V(e);
} : la, Xu = 200;
function Pn(e, t, n) {
  var r = -1, a = Ma, i = e.length, o = !0, s = [], u = s;
  if (n)
    o = !1, a = Nu;
  else if (i >= Xu) {
    var c = t ? null : Qu(e);
    if (c)
      return ot(c);
    o = !1, a = vn, u = new fe();
  } else
    u = t ? [] : s;
  e:
    for (; ++r < i; ) {
      var l = e[r], f = t ? t(l) : l;
      if (l = n || l !== 0 ? l : 0, o && f === f) {
        for (var d = u.length; d--; )
          if (u[d] === f)
            continue e;
        t && u.push(f), s.push(l);
      } else
        a(u, f, n) || (u !== s && u.push(f), s.push(l));
    }
  return s;
}
var ku = Jt(function(e) {
  var t = Wu(e);
  return qe(t) && (t = void 0), Pn(cn(e, 1, qe, !0), Iu(t));
});
function bl(e) {
  return e && e.length ? Pn(e) : [];
}
const Vu = "cds", Zu = {
  opacity: {
    unselected: 0.05,
    selected: 0.4
  }
}, Ku = {
  ticks: {
    number: 7,
    rotateIfSmallerThan: 30,
    verticalSpaceRatio: 2.5,
    horizontalSpaceRatio: 3.5
  },
  ratio: {
    reference: "value",
    compareTo: "marker"
  },
  paddingRatio: 0.1,
  hover: {
    rectanglePadding: 4
  }
}, Ju = {
  duration: 1e3,
  ease: kn,
  zoomLevel: 3
}, xn = {
  circles: {
    fillOpacity: 0.3,
    hover: {
      stroke: "#FFF"
    }
  },
  padding: {
    mainGroup: 4,
    children: 2
  },
  hierarchyLevel: 2
}, ec = {
  pairingOptions: {
    "1-color": 4,
    "2-color": 5,
    "3-color": 5,
    "4-color": 3,
    "5-color": 2,
    "14-color": 1
  }
}, tc = {
  circle: {
    radius: 4,
    opacity: {
      hovered: 1,
      default: 0.3
    }
  },
  box: {
    opacity: {
      hovered: 0.5,
      default: 0.3
    }
  },
  strokeWidth: {
    default: 1,
    thicker: 2
  }
}, nc = {
  defaultBins: 10
}, rc = {
  items: {
    status: {
      ACTIVE: 1,
      DISABLED: 0
    },
    horizontalSpace: 12,
    verticalSpace: 24,
    textYOffset: 8,
    spaceAfter: 4
  },
  checkbox: {
    radius: 6.5
  },
  radius: {
    iconData: [
      { cx: 7, cy: 7, r: 6.5 },
      { cx: 7, cy: 10, r: 3.5 }
    ],
    fill: null,
    stroke: "#8c8c8c"
  },
  line: {
    yPosition: 6,
    width: 24,
    strokeWidth: 1.4,
    fill: null,
    stroke: "#999999"
  },
  area: {
    width: 24,
    height: 14,
    fill: "#6f6f6f",
    stroke: null
  },
  size: {
    iconData: [
      { width: 23, height: 12 },
      { width: 13, height: 6 }
    ],
    fill: null,
    stroke: "#8D8D8D"
  },
  quartile: {
    iconData: [
      { x: 0, y: 0, width: 24, height: 13 },
      { x: 11, y: 4, width: 1, height: 4 }
    ]
  },
  zoom: {
    iconData: [{ x: 0, y: 0, width: 12, height: 12 }],
    color: "#8D8D8D"
  },
  color: {
    barWidth: 300,
    barHeight: 8,
    axisYTranslation: 10
  }
}, ac = {
  opacity: {
    unselected: 0.3,
    selected: 1
  },
  weight: {
    selected: 2,
    unselected: 1
  }
}, ic = {
  statusBar: {
    paddingRight: 5
  },
  status: {
    indicatorSize: 16,
    paddingLeft: 15
  },
  total: {
    paddingLeft: 36,
    paddingRight: 24
  },
  height: {
    default: 8,
    proportional: 16
  },
  dividerWidth: 2
}, oc = {
  radiusOffset: -15,
  innerRadius: 2,
  padAngle: 7e-3,
  hoverArc: {
    outerRadiusOffset: 3
  },
  xOffset: 30,
  yOffset: 20,
  yOffsetCallout: 10,
  callout: {
    minSliceDegree: 5,
    offsetX: 15,
    offsetY: 12,
    horizontalLineLength: 8,
    textMargin: 2
  }
}, sc = {
  opacity: {
    unselected: 0.1,
    selected: 0.3
  },
  xLabelPadding: 10,
  yLabelPadding: 8,
  yTicksNumber: 4,
  minRange: 10,
  xAxisRectHeight: 50,
  dotsRadius: 5
}, uc = {
  nodeWidth: 4,
  nodeHoveredWidth: 8,
  minNodePadding: 24,
  opacity: {
    unfocus: 0.3,
    default: 0.8,
    selected: 1
  }
}, cc = {
  minCellDividerDimension: 16,
  // Ensures axes lines are displayed with or without stroke disabled
  chartPadding: 0.5
}, lc = {
  default: {
    size: 24
  }
}, fc = {
  defaultOffset: 4,
  horizontalOffset: 10
}, dc = {
  default: {
    duration: 300
  },
  pie_slice_mouseover: {
    duration: 100
  },
  pie_slice_mouseout: {
    duration: 100
  },
  pie_chart_titles: {
    duration: 375
  },
  graph_element_mouseover_fill_update: {
    duration: 100
  },
  graph_element_mouseout_fill_update: {
    duration: 100
  }
}, hc = {
  buttonSize: 32,
  iconSize: 20,
  height: 32,
  spacerHeight: 10,
  iconPadding: 6
}, gc = {
  height: {
    [Ne.GRAPH_VIEW]: 32,
    [Ne.SLIDER_VIEW]: 10
  },
  spacerHeight: 8,
  handleWidth: 5,
  handleBarWidth: 1,
  handleBarHeight: 12
}, An = [
  {
    type: $.RADIUS,
    name: "Radius"
  },
  {
    type: $.AREA,
    name: "Poor area"
  },
  {
    type: $.AREA,
    name: "Satisfactory area"
  },
  {
    type: $.AREA,
    name: "Great area"
  },
  {
    type: $.QUARTILE,
    name: "Quartiles"
  }
], k = {
  type: Te.END_LINE,
  threshold: 16,
  numCharacter: 14
}, mc = {
  code: typeof navigator < "u" && (navigator == null ? void 0 : navigator.language) || "en-US",
  // read from browser's navigator.language
  number: (e, t = (navigator == null ? void 0 : navigator.language) || "en-US") => e.toLocaleString(t),
  // based on code property if specified
  date: (e, t = (navigator == null ? void 0 : navigator.language) || "en-US", n = {}, r = null) => r || e.toLocaleDateString(t, n),
  // based on code property if specified
  time: (e, t = (navigator == null ? void 0 : navigator.language) || "en-US", n = {}, r = null) => r || e.toLocaleTimeString(t, n),
  // based on code property if specified
  optionsObject: {
    "15seconds": {
      primary: {
        "MMM d, pp": {
          month: "short",
          day: "numeric",
          hour: "numeric",
          minute: "2-digit",
          second: "2-digit",
          hourCycle: "h12"
        },
        "MMM d, h:mm:ss.SSS a": {
          month: "short",
          day: "numeric",
          hour: "numeric",
          minute: "2-digit",
          fractionalSecondDigits: 3,
          hourCycle: "h12"
        }
      },
      secondary: {
        pp: {
          hour: "numeric",
          minute: "2-digit",
          second: "2-digit",
          hourCycle: "h12"
        },
        "h:mm:ss.SSS a": {
          hour: "numeric",
          minute: "2-digit",
          fractionalSecondDigits: 3,
          hourCycle: "h12"
        }
      },
      type: "time"
    },
    minute: {
      primary: {
        "MMM d, p": {
          month: "short",
          day: "numeric",
          hour: "numeric",
          minute: "2-digit",
          hourCycle: "h12"
        }
      },
      secondary: {
        p: {
          hour: "numeric",
          minute: "2-digit",
          hourCycle: "h12"
        }
      },
      type: "time"
    },
    "30minutes": {
      primary: {
        "MMM d, p": {
          month: "short",
          day: "numeric",
          hour: "numeric",
          minute: "2-digit",
          hourCycle: "h12"
        }
      },
      secondary: {
        p: {
          hour: "numeric",
          minute: "2-digit",
          hourCycle: "h12"
        }
      },
      type: "time"
    },
    hourly: {
      primary: {
        "MMM d, hh a": {
          month: "short",
          day: "numeric",
          hour: "2-digit",
          hourCycle: "h12"
        }
      },
      secondary: {
        "hh a": {
          hour: "2-digit",
          hourCycle: "h12"
        }
      },
      type: "time"
    },
    daily: {
      primary: {
        "MMM d": {
          month: "short",
          day: "numeric"
        }
      },
      secondary: {
        d: {
          day: "numeric"
        }
      },
      type: "date"
    },
    weekly: {
      primary: {
        "eee, MMM d": {
          weekday: "short",
          month: "short",
          day: "numeric"
        }
      },
      secondary: {
        eee: {
          weekday: "short"
        }
      },
      type: "date"
    },
    monthly: {
      primary: {
        "MMM yyyy": {
          month: "short",
          year: "numeric"
        }
      },
      secondary: {
        MMM: {
          month: "short"
        }
      },
      type: "date"
    },
    quarterly: {
      primary: {},
      secondary: {},
      type: "date"
    },
    yearly: {
      primary: {
        yyyy: {
          year: "numeric"
        }
      },
      secondary: {
        yyyy: {
          year: "numeric"
        }
      },
      type: "date"
    }
  },
  translations: {
    group: "Group",
    total: "Total",
    meter: {
      title: ""
      //default is emply string as meter title is dataset label
    },
    tabularRep: {
      title: "Tabular representation",
      downloadAsCSV: "Download as CSV"
    },
    toolbar: {
      exportAsCSV: "Export to CSV",
      exportAsJPG: "Export to JPG",
      exportAsPNG: "Export to PNG",
      zoomIn: "Zoom in",
      zoomOut: "Zoom out",
      resetZoom: "Reset zoom",
      moreOptions: "More options",
      makeFullScreen: "Make fullscreen",
      exitFullScreen: "Exit fullscreen",
      showAsTable: "Show as table"
    }
  }
}, yc = {
  enabled: !0,
  position: Hn.BOTTOM,
  clickable: !0,
  truncation: k,
  alignment: te.LEFT,
  order: null,
  additionalItems: []
}, En = {
  x: {
    // set enable to false will not draw grid and stroke of grid backdrop
    enabled: !0,
    numberOfTicks: 15,
    alignWithAxisTicks: !1
  },
  y: {
    // set enable to false will not draw grid and stroke of grid backdrop
    enabled: !0,
    numberOfTicks: 5,
    alignWithAxisTicks: !1
  }
}, $n = {
  // enable or disable ruler
  enabled: !0
}, ut = {
  enabled: !0,
  showTotal: !0,
  truncation: k,
  groupLabel: "Group"
}, Cn = {
  top: {
    visible: !0,
    includeZero: !0,
    truncation: k
  },
  bottom: {
    visible: !0,
    includeZero: !0,
    truncation: k
  },
  left: {
    visible: !0,
    includeZero: !0,
    truncation: k
  },
  right: {
    visible: !0,
    includeZero: !0,
    truncation: k
  }
}, ye = {
  addSpaceOnEdges: 1,
  showDayName: !1,
  localeObject: Qt,
  timeIntervalFormats: {
    "15seconds": { primary: "MMM d, pp", secondary: "pp" },
    minute: { primary: "MMM d, p", secondary: "p" },
    "30minutes": { primary: "MMM d, p", secondary: "p" },
    hourly: { primary: "MMM d, hh a", secondary: "hh a" },
    daily: { primary: "MMM d", secondary: "d" },
    weekly: { primary: "eee, MMM d", secondary: "eee" },
    monthly: { primary: "MMM yyyy", secondary: "MMM" },
    quarterly: { primary: "QQQ ''yy", secondary: "QQQ" },
    yearly: { primary: "yyyy", secondary: "yyyy" }
  }
};
let Dn = !1;
try {
  Dn = typeof document < "u" && (document.fullscreenEnabled || document.webkitFullscreenEnabled || document.mozFullScreenEnabled || document.msFullscreenEnabled);
} catch (e) {
  console.warn("Fullscreen capabilities check failed: ", e.message);
}
const M = {
  width: null,
  height: null,
  resizable: !0,
  theme: qn.WHITE,
  tooltip: ut,
  legend: yc,
  locale: mc,
  style: {
    prefix: "cc"
  },
  data: {
    groupMapsTo: "group",
    loading: !1,
    selectedGroups: []
  },
  color: {
    scale: null,
    pairing: {
      numberOfVariants: null,
      option: 1
    },
    gradient: {
      enabled: !1
    }
  },
  toolbar: {
    enabled: !0,
    numberOfIcons: 3,
    controls: [
      {
        type: ne.SHOW_AS_DATATABLE
      },
      ...Dn ? [
        {
          type: ne.MAKE_FULLSCREEN
        }
      ] : [],
      {
        type: ne.EXPORT_CSV
      },
      {
        type: ne.EXPORT_PNG
      },
      {
        type: ne.EXPORT_JPG
      }
    ]
  }
}, In = h({}, M, {
  thematic: {
    projection: Un.geoNaturalEarth1
  }
}), pe = h({}, M, {
  axes: Cn,
  timeScale: ye,
  grid: En,
  ruler: $n,
  zoomBar: {
    zoomRatio: 0.4,
    minZoomRatio: 0.01,
    top: {
      enabled: !1,
      type: Ne.GRAPH_VIEW
    }
  }
}), H = h({}, pe, {
  bars: {
    maxWidth: 16,
    spacingFactor: 0.25
  },
  timeScale: h(ye, {
    addSpaceOnEdges: 1
  })
}), pc = h({}, H, {}), bc = h({}, H, {}), wc = h({}, H, {
  bars: h({}, H.bars, {
    dividerSize: 1.5
  })
}), Tc = h({}, H, {}), ct = h({}, pe, {
  points: {
    // default point radius to 4
    radius: 4,
    fillOpacity: 0.3,
    filled: !0,
    enabled: !0
  }
}), vc = ct, Fn = h({}, ct, {
  points: {
    // default point radius to 3
    radius: 3,
    filled: !1,
    enabled: !0
  }
}), Ln = h({}, Fn, {
  timeScale: h(ye, {
    addSpaceOnEdges: 0
  })
}), Oc = Ln, Sc = h({}, pe, {
  bubble: {
    radiusMapsTo: "radius",
    radiusLabel: "Radius",
    radiusRange: (e) => {
      const t = Math.min(e.width, e.height);
      return [t * 3 / 400, t * 25 / 400];
    },
    fillOpacity: 0.2,
    enabled: !0
  },
  points: {
    filled: !0
  },
  legend: {
    additionalItems: [
      {
        type: $.RADIUS,
        name: "Radius"
      }
    ]
  }
}), Mc = h({}, pe, {
  bullet: {
    performanceAreaTitles: ["Poor", "Satisfactory", "Great"]
  },
  grid: {
    x: {
      enabled: !1
    },
    y: {
      enabled: !1
    }
  },
  legend: {
    additionalItems: [
      {
        type: $.AREA,
        name: "Poor area"
      },
      {
        type: $.AREA,
        name: "Satisfactory area"
      },
      {
        type: $.AREA,
        name: "Great area"
      },
      {
        type: $.QUARTILE,
        name: "Quartiles"
      }
    ]
  }
}), _c = h({}, H, {
  bars: {
    dividerSize: 1.5
  },
  timeScale: h(ye, {
    addSpaceOnEdges: 0
  })
}), Pc = h({}, M, {
  tooltip: h({}, ut, {
    wordLabel: "Word",
    valueLabel: "Value"
  }),
  wordCloud: {
    fontSizeMapsTo: "value",
    fontSizeRange: (e) => {
      const t = Math.min(e.width, e.height);
      return [t * 20 / 400, t * 75 / 400];
    },
    wordMapsTo: "word"
  }
}), Rn = h({}, M, {
  pie: {
    labels: {
      formatter: null,
      enabled: !0
    },
    alignment: te.LEFT,
    sortFunction: null,
    valueMapsTo: "value"
  }
}), xc = h({}, M, {
  legend: {
    enabled: !1
  },
  gauge: {
    type: zn.SEMI,
    arcWidth: 16,
    deltaArrow: {
      size: (e) => e / 8,
      enabled: !0
    },
    showPercentageSymbol: !0,
    status: null,
    numberSpacing: 10,
    deltaFontSize: (e) => e / 8,
    valueFontSize: (e) => e / 2.5,
    alignment: te.LEFT
  }
}), Ac = h({}, Rn, {
  donut: {
    center: {
      numberFontSize: (e) => `${Math.min(e / 100 * 24, 24)}px`,
      titleFontSize: (e) => `${Math.min(e / 100 * 15, 15)}px`,
      titleYPosition: (e) => Math.min(e / 80 * 20, 20)
    },
    alignment: te.LEFT
  }
}), jn = h({}, M, {
  legend: {
    enabled: !1,
    clickable: !1
  },
  meter: {
    showLabels: !0,
    proportional: null,
    statusBar: {
      percentageIndicator: {
        enabled: !0
      }
    }
  }
}), Ec = h({}, jn, {
  legend: {
    enabled: !0
  }
}), $c = h({}, M, {
  radar: {
    axes: {
      angle: "key",
      value: "value"
    },
    alignment: te.LEFT
  },
  tooltip: {
    gridline: {
      enabled: !0
    }
  }
}), Cc = h({}, H, {
  comboChartTypes: []
}), Dc = h(
  {
    tree: {
      type: Qn.TREE
    }
  },
  M,
  {}
), Ic = h({}, M, {
  data: h(M.data, {
    groupMapsTo: "name"
  })
}), Fc = h({}, M, xn, {
  data: h(M.data, {
    groupMapsTo: "name"
  })
}), Lc = h({}, M, {
  alluvial: {
    data: h(M.data, {
      groupMapsTo: "source"
    }),
    nodeAlignment: te.CENTER,
    nodePadding: 24,
    monochrome: !1,
    nodes: []
  }
}), Rc = h({}, M, {
  axes: Cn,
  heatmap: {
    divider: {
      state: Xn.AUTO
    },
    colorLegend: {
      type: "linear"
    }
  }
}), jc = h({}, In, {
  choropleth: {
    colorLegend: {
      type: "linear"
    }
  }
}), Nc = {
  alluvialChart: Lc,
  areaChart: Ln,
  axisChart: pe,
  boxplotChart: Tc,
  bubbleChart: Sc,
  bulletChart: Mc,
  chart: M,
  circlePackChart: Fc,
  choroplethChart: jc,
  comboChart: Cc,
  donutChart: Ac,
  gaugeChart: xc,
  groupedBarChart: bc,
  heatmapChart: Rc,
  histogramChart: _c,
  lineChart: Fn,
  lollipopChart: vc,
  meterChart: jn,
  pieChart: Rn,
  proportionalMeterChart: Ec,
  radarChart: $c,
  scatterChart: ct,
  simpleBarChart: pc,
  stackedAreaChart: Oc,
  stackedBarChart: wc,
  thematicChart: In,
  treeChart: Dc,
  treemapChart: Ic,
  wordCloudChart: Pc
}, wl = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  alluvial: uc,
  area: Zu,
  axis: Ku,
  baseTooltip: ut,
  boxplot: tc,
  canvasZoomSettings: Ju,
  carbonPrefix: Vu,
  circlePack: xn,
  color: ec,
  defaultLegendAdditionalItems: An,
  grid: En,
  heatmap: cc,
  histogram: nc,
  legend: rc,
  lines: ac,
  meter: ic,
  options: Nc,
  pie: oc,
  radar: sc,
  ruler: $n,
  spacers: lc,
  timeScale: ye,
  toolbar: hc,
  tooltips: fc,
  transitions: dc,
  zoomBar: gc
}, Symbol.toStringTag, { value: "Module" }));
function Tl(e, t, n) {
  let r = null;
  return function(...a) {
    const i = this;
    i.mousePosition = Vn(a[0], n), clearTimeout(r), r = setTimeout(function() {
      e.apply(i, a);
    }, t);
  };
}
function vl(e, t) {
  var a;
  const n = qs(e), r = Object.keys(t.axes || {});
  (a = t == null ? void 0 : t.toolbar) != null && a.controls && delete n.toolbar.controls, r.length === 0 && delete n.axes;
  for (const i in n.axes)
    if (r.includes(i)) {
      const o = t.axes[i];
      (o.primary || o.secondary) && console.warn(
        "`primary` & `secondary` are no longer needed for axis configurations. Read more here https://charts.carbondesignsystem.com/?path=/story/docs-tutorials--tabular-data-format"
      );
      const s = o.mapsTo;
      if (s == null) {
        const u = o.scaleType;
        u == null ? o.mapsTo = "value" : u === ft.TIME ? o.mapsTo = "date" : u === ft.LABELS && (o.mapsTo = "key");
      }
    } else
      delete n.axes[i];
  return Wc(n, t), h(n, t);
}
function Ol(e) {
  if (!e)
    return;
  const t = /translate\([0-9]+\.?[0-9]*,[0-9]+\.?[0-9]*\)/, n = e.getAttribute("transform").match(t);
  if (!n)
    return null;
  if (n[0]) {
    const r = n[0].replace(/translate\(/, "").replace(/\)/, "").split(",");
    return {
      tx: r[0],
      ty: r[1]
    };
  }
  return null;
}
function Sl(e) {
  const n = /\(([^)]+)\)/.exec(e);
  if (n && n.length > 1) {
    const r = n[1].split(",");
    if (r.length > 1)
      return {
        x: parseFloat(r[0]),
        y: parseFloat(r[1])
      };
  }
  return { x: 0, y: 0 };
}
function Ml(e, t, n = "value", r = !1) {
  const a = e / t.reduce((i, o) => i + o[n], 0) * 100;
  return r ? a : a % 1 !== 0 ? parseFloat(a.toFixed(1)) : a;
}
function _l(e, t, n) {
  if (n > e.length)
    return e;
  if (t === Te.MID_LINE)
    return e.substr(0, n / 2) + "..." + e.substr(-n / 2);
  if (t === Te.FRONT_LINE)
    return "..." + e.substr(-n);
  if (t === Te.END_LINE)
    return e.substr(0, n) + "...";
}
function Wc(e, t) {
  const n = G(e, "legend", "additionalItems"), r = G(t, "legend", "additionalItems");
  if (n && r) {
    const a = r.map((s) => s.type), i = n.map((s) => s.type), o = An.filter(
      (s) => i.includes(s.type) && !a.includes(s.type)
    );
    e.legend.additionalItems = o, t.legend.additionalItems = ku(
      o,
      r,
      "name"
    );
  }
}
const G = (e, ...t) => {
  let n = e;
  if (n) {
    for (const r of t)
      if (n[r] !== null && n[r] !== void 0)
        n = n[r];
      else
        return null;
    return n;
  }
  return null;
}, Yc = (e, t) => t === zt.HORIZONTAL ? {
  y0: e.x0,
  y1: e.x1,
  x0: e.y0,
  x1: e.y1
} : e, Pl = (e, t) => {
  const { x0: n, x1: r, y0: a, y1: i } = Yc(e, t);
  return `M${n},${a}L${n},${i}L${r},${i}L${r},${a}L${n},${a}`;
};
function xl(e, t, n) {
  return n === zt.VERTICAL ? [e, t] : [t, e];
}
const Nn = 6048e5, Gc = 864e5, Al = 6e4, El = 36e5;
function Yt(e) {
  const t = A(e);
  return t.setHours(0, 0, 0, 0), t;
}
function Gt(e) {
  const t = A(e), n = new Date(
    Date.UTC(
      t.getFullYear(),
      t.getMonth(),
      t.getDate(),
      t.getHours(),
      t.getMinutes(),
      t.getSeconds(),
      t.getMilliseconds()
    )
  );
  return n.setUTCFullYear(t.getFullYear()), +e - +n;
}
function Bc(e, t) {
  const n = Yt(e), r = Yt(t), a = +n - Gt(n), i = +r - Gt(r);
  return Math.round((a - i) / Gc);
}
function q(e, t) {
  return e instanceof Date ? new e.constructor(t) : new Date(t);
}
function Hc(e) {
  const t = A(e), n = q(e, 0);
  return n.setFullYear(t.getFullYear(), 0, 1), n.setHours(0, 0, 0, 0), n;
}
function qc(e) {
  const t = A(e);
  return Bc(t, Hc(t)) + 1;
}
function Me(e) {
  return ue(e, { weekStartsOn: 1 });
}
function Wn(e) {
  const t = A(e), n = t.getFullYear(), r = q(e, 0);
  r.setFullYear(n + 1, 0, 4), r.setHours(0, 0, 0, 0);
  const a = Me(r), i = q(e, 0);
  i.setFullYear(n, 0, 4), i.setHours(0, 0, 0, 0);
  const o = Me(i);
  return t.getTime() >= a.getTime() ? n + 1 : t.getTime() >= o.getTime() ? n : n - 1;
}
function Uc(e) {
  const t = Wn(e), n = q(e, 0);
  return n.setFullYear(t, 0, 4), n.setHours(0, 0, 0, 0), Me(n);
}
function zc(e) {
  const t = A(e), n = +Me(t) - +Uc(t);
  return Math.round(n / Nn) + 1;
}
function Yn(e, t) {
  var l, f, d, m;
  const n = A(e), r = n.getFullYear(), a = Pe(), i = (t == null ? void 0 : t.firstWeekContainsDate) ?? ((f = (l = t == null ? void 0 : t.locale) == null ? void 0 : l.options) == null ? void 0 : f.firstWeekContainsDate) ?? a.firstWeekContainsDate ?? ((m = (d = a.locale) == null ? void 0 : d.options) == null ? void 0 : m.firstWeekContainsDate) ?? 1, o = q(e, 0);
  o.setFullYear(r + 1, 0, i), o.setHours(0, 0, 0, 0);
  const s = ue(o, t), u = q(e, 0);
  u.setFullYear(r, 0, i), u.setHours(0, 0, 0, 0);
  const c = ue(u, t);
  return n.getTime() >= s.getTime() ? r + 1 : n.getTime() >= c.getTime() ? r : r - 1;
}
function Qc(e, t) {
  var s, u, c, l;
  const n = Pe(), r = (t == null ? void 0 : t.firstWeekContainsDate) ?? ((u = (s = t == null ? void 0 : t.locale) == null ? void 0 : s.options) == null ? void 0 : u.firstWeekContainsDate) ?? n.firstWeekContainsDate ?? ((l = (c = n.locale) == null ? void 0 : c.options) == null ? void 0 : l.firstWeekContainsDate) ?? 1, a = Yn(e, t), i = q(e, 0);
  return i.setFullYear(a, 0, r), i.setHours(0, 0, 0, 0), ue(i, t);
}
function Xc(e, t) {
  const n = A(e), r = +ue(n, t) - +Qc(n, t);
  return Math.round(r / Nn) + 1;
}
function g(e, t) {
  const n = e < 0 ? "-" : "", r = Math.abs(e).toString().padStart(t, "0");
  return n + r;
}
const j = {
  // Year
  y(e, t) {
    const n = e.getFullYear(), r = n > 0 ? n : 1 - n;
    return g(t === "yy" ? r % 100 : r, t.length);
  },
  // Month
  M(e, t) {
    const n = e.getMonth();
    return t === "M" ? String(n + 1) : g(n + 1, 2);
  },
  // Day of the month
  d(e, t) {
    return g(e.getDate(), t.length);
  },
  // AM or PM
  a(e, t) {
    const n = e.getHours() / 12 >= 1 ? "pm" : "am";
    switch (t) {
      case "a":
      case "aa":
        return n.toUpperCase();
      case "aaa":
        return n;
      case "aaaaa":
        return n[0];
      case "aaaa":
      default:
        return n === "am" ? "a.m." : "p.m.";
    }
  },
  // Hour [1-12]
  h(e, t) {
    return g(e.getHours() % 12 || 12, t.length);
  },
  // Hour [0-23]
  H(e, t) {
    return g(e.getHours(), t.length);
  },
  // Minute
  m(e, t) {
    return g(e.getMinutes(), t.length);
  },
  // Second
  s(e, t) {
    return g(e.getSeconds(), t.length);
  },
  // Fraction of second
  S(e, t) {
    const n = t.length, r = e.getMilliseconds(), a = Math.trunc(
      r * Math.pow(10, n - 3)
    );
    return g(a, t.length);
  }
}, X = {
  am: "am",
  pm: "pm",
  midnight: "midnight",
  noon: "noon",
  morning: "morning",
  afternoon: "afternoon",
  evening: "evening",
  night: "night"
}, Bt = {
  // Era
  G: function(e, t, n) {
    const r = e.getFullYear() > 0 ? 1 : 0;
    switch (t) {
      case "G":
      case "GG":
      case "GGG":
        return n.era(r, { width: "abbreviated" });
      case "GGGGG":
        return n.era(r, { width: "narrow" });
      case "GGGG":
      default:
        return n.era(r, { width: "wide" });
    }
  },
  // Year
  y: function(e, t, n) {
    if (t === "yo") {
      const r = e.getFullYear(), a = r > 0 ? r : 1 - r;
      return n.ordinalNumber(a, { unit: "year" });
    }
    return j.y(e, t);
  },
  // Local week-numbering year
  Y: function(e, t, n, r) {
    const a = Yn(e, r), i = a > 0 ? a : 1 - a;
    if (t === "YY") {
      const o = i % 100;
      return g(o, 2);
    }
    return t === "Yo" ? n.ordinalNumber(i, { unit: "year" }) : g(i, t.length);
  },
  // ISO week-numbering year
  R: function(e, t) {
    const n = Wn(e);
    return g(n, t.length);
  },
  // Extended year. This is a single number designating the year of this calendar system.
  // The main difference between `y` and `u` localizers are B.C. years:
  // | Year | `y` | `u` |
  // |------|-----|-----|
  // | AC 1 |   1 |   1 |
  // | BC 1 |   1 |   0 |
  // | BC 2 |   2 |  -1 |
  // Also `yy` always returns the last two digits of a year,
  // while `uu` pads single digit years to 2 characters and returns other years unchanged.
  u: function(e, t) {
    const n = e.getFullYear();
    return g(n, t.length);
  },
  // Quarter
  Q: function(e, t, n) {
    const r = Math.ceil((e.getMonth() + 1) / 3);
    switch (t) {
      case "Q":
        return String(r);
      case "QQ":
        return g(r, 2);
      case "Qo":
        return n.ordinalNumber(r, { unit: "quarter" });
      case "QQQ":
        return n.quarter(r, {
          width: "abbreviated",
          context: "formatting"
        });
      case "QQQQQ":
        return n.quarter(r, {
          width: "narrow",
          context: "formatting"
        });
      case "QQQQ":
      default:
        return n.quarter(r, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // Stand-alone quarter
  q: function(e, t, n) {
    const r = Math.ceil((e.getMonth() + 1) / 3);
    switch (t) {
      case "q":
        return String(r);
      case "qq":
        return g(r, 2);
      case "qo":
        return n.ordinalNumber(r, { unit: "quarter" });
      case "qqq":
        return n.quarter(r, {
          width: "abbreviated",
          context: "standalone"
        });
      case "qqqqq":
        return n.quarter(r, {
          width: "narrow",
          context: "standalone"
        });
      case "qqqq":
      default:
        return n.quarter(r, {
          width: "wide",
          context: "standalone"
        });
    }
  },
  // Month
  M: function(e, t, n) {
    const r = e.getMonth();
    switch (t) {
      case "M":
      case "MM":
        return j.M(e, t);
      case "Mo":
        return n.ordinalNumber(r + 1, { unit: "month" });
      case "MMM":
        return n.month(r, {
          width: "abbreviated",
          context: "formatting"
        });
      case "MMMMM":
        return n.month(r, {
          width: "narrow",
          context: "formatting"
        });
      case "MMMM":
      default:
        return n.month(r, { width: "wide", context: "formatting" });
    }
  },
  // Stand-alone month
  L: function(e, t, n) {
    const r = e.getMonth();
    switch (t) {
      case "L":
        return String(r + 1);
      case "LL":
        return g(r + 1, 2);
      case "Lo":
        return n.ordinalNumber(r + 1, { unit: "month" });
      case "LLL":
        return n.month(r, {
          width: "abbreviated",
          context: "standalone"
        });
      case "LLLLL":
        return n.month(r, {
          width: "narrow",
          context: "standalone"
        });
      case "LLLL":
      default:
        return n.month(r, { width: "wide", context: "standalone" });
    }
  },
  // Local week of year
  w: function(e, t, n, r) {
    const a = Xc(e, r);
    return t === "wo" ? n.ordinalNumber(a, { unit: "week" }) : g(a, t.length);
  },
  // ISO week of year
  I: function(e, t, n) {
    const r = zc(e);
    return t === "Io" ? n.ordinalNumber(r, { unit: "week" }) : g(r, t.length);
  },
  // Day of the month
  d: function(e, t, n) {
    return t === "do" ? n.ordinalNumber(e.getDate(), { unit: "date" }) : j.d(e, t);
  },
  // Day of year
  D: function(e, t, n) {
    const r = qc(e);
    return t === "Do" ? n.ordinalNumber(r, { unit: "dayOfYear" }) : g(r, t.length);
  },
  // Day of week
  E: function(e, t, n) {
    const r = e.getDay();
    switch (t) {
      case "E":
      case "EE":
      case "EEE":
        return n.day(r, {
          width: "abbreviated",
          context: "formatting"
        });
      case "EEEEE":
        return n.day(r, {
          width: "narrow",
          context: "formatting"
        });
      case "EEEEEE":
        return n.day(r, {
          width: "short",
          context: "formatting"
        });
      case "EEEE":
      default:
        return n.day(r, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // Local day of week
  e: function(e, t, n, r) {
    const a = e.getDay(), i = (a - r.weekStartsOn + 8) % 7 || 7;
    switch (t) {
      case "e":
        return String(i);
      case "ee":
        return g(i, 2);
      case "eo":
        return n.ordinalNumber(i, { unit: "day" });
      case "eee":
        return n.day(a, {
          width: "abbreviated",
          context: "formatting"
        });
      case "eeeee":
        return n.day(a, {
          width: "narrow",
          context: "formatting"
        });
      case "eeeeee":
        return n.day(a, {
          width: "short",
          context: "formatting"
        });
      case "eeee":
      default:
        return n.day(a, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // Stand-alone local day of week
  c: function(e, t, n, r) {
    const a = e.getDay(), i = (a - r.weekStartsOn + 8) % 7 || 7;
    switch (t) {
      case "c":
        return String(i);
      case "cc":
        return g(i, t.length);
      case "co":
        return n.ordinalNumber(i, { unit: "day" });
      case "ccc":
        return n.day(a, {
          width: "abbreviated",
          context: "standalone"
        });
      case "ccccc":
        return n.day(a, {
          width: "narrow",
          context: "standalone"
        });
      case "cccccc":
        return n.day(a, {
          width: "short",
          context: "standalone"
        });
      case "cccc":
      default:
        return n.day(a, {
          width: "wide",
          context: "standalone"
        });
    }
  },
  // ISO day of week
  i: function(e, t, n) {
    const r = e.getDay(), a = r === 0 ? 7 : r;
    switch (t) {
      case "i":
        return String(a);
      case "ii":
        return g(a, t.length);
      case "io":
        return n.ordinalNumber(a, { unit: "day" });
      case "iii":
        return n.day(r, {
          width: "abbreviated",
          context: "formatting"
        });
      case "iiiii":
        return n.day(r, {
          width: "narrow",
          context: "formatting"
        });
      case "iiiiii":
        return n.day(r, {
          width: "short",
          context: "formatting"
        });
      case "iiii":
      default:
        return n.day(r, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // AM or PM
  a: function(e, t, n) {
    const a = e.getHours() / 12 >= 1 ? "pm" : "am";
    switch (t) {
      case "a":
      case "aa":
        return n.dayPeriod(a, {
          width: "abbreviated",
          context: "formatting"
        });
      case "aaa":
        return n.dayPeriod(a, {
          width: "abbreviated",
          context: "formatting"
        }).toLowerCase();
      case "aaaaa":
        return n.dayPeriod(a, {
          width: "narrow",
          context: "formatting"
        });
      case "aaaa":
      default:
        return n.dayPeriod(a, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // AM, PM, midnight, noon
  b: function(e, t, n) {
    const r = e.getHours();
    let a;
    switch (r === 12 ? a = X.noon : r === 0 ? a = X.midnight : a = r / 12 >= 1 ? "pm" : "am", t) {
      case "b":
      case "bb":
        return n.dayPeriod(a, {
          width: "abbreviated",
          context: "formatting"
        });
      case "bbb":
        return n.dayPeriod(a, {
          width: "abbreviated",
          context: "formatting"
        }).toLowerCase();
      case "bbbbb":
        return n.dayPeriod(a, {
          width: "narrow",
          context: "formatting"
        });
      case "bbbb":
      default:
        return n.dayPeriod(a, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // in the morning, in the afternoon, in the evening, at night
  B: function(e, t, n) {
    const r = e.getHours();
    let a;
    switch (r >= 17 ? a = X.evening : r >= 12 ? a = X.afternoon : r >= 4 ? a = X.morning : a = X.night, t) {
      case "B":
      case "BB":
      case "BBB":
        return n.dayPeriod(a, {
          width: "abbreviated",
          context: "formatting"
        });
      case "BBBBB":
        return n.dayPeriod(a, {
          width: "narrow",
          context: "formatting"
        });
      case "BBBB":
      default:
        return n.dayPeriod(a, {
          width: "wide",
          context: "formatting"
        });
    }
  },
  // Hour [1-12]
  h: function(e, t, n) {
    if (t === "ho") {
      let r = e.getHours() % 12;
      return r === 0 && (r = 12), n.ordinalNumber(r, { unit: "hour" });
    }
    return j.h(e, t);
  },
  // Hour [0-23]
  H: function(e, t, n) {
    return t === "Ho" ? n.ordinalNumber(e.getHours(), { unit: "hour" }) : j.H(e, t);
  },
  // Hour [0-11]
  K: function(e, t, n) {
    const r = e.getHours() % 12;
    return t === "Ko" ? n.ordinalNumber(r, { unit: "hour" }) : g(r, t.length);
  },
  // Hour [1-24]
  k: function(e, t, n) {
    let r = e.getHours();
    return r === 0 && (r = 24), t === "ko" ? n.ordinalNumber(r, { unit: "hour" }) : g(r, t.length);
  },
  // Minute
  m: function(e, t, n) {
    return t === "mo" ? n.ordinalNumber(e.getMinutes(), { unit: "minute" }) : j.m(e, t);
  },
  // Second
  s: function(e, t, n) {
    return t === "so" ? n.ordinalNumber(e.getSeconds(), { unit: "second" }) : j.s(e, t);
  },
  // Fraction of second
  S: function(e, t) {
    return j.S(e, t);
  },
  // Timezone (ISO-8601. If offset is 0, output is always `'Z'`)
  X: function(e, t, n) {
    const r = e.getTimezoneOffset();
    if (r === 0)
      return "Z";
    switch (t) {
      case "X":
        return qt(r);
      case "XXXX":
      case "XX":
        return Y(r);
      case "XXXXX":
      case "XXX":
      default:
        return Y(r, ":");
    }
  },
  // Timezone (ISO-8601. If offset is 0, output is `'+00:00'` or equivalent)
  x: function(e, t, n) {
    const r = e.getTimezoneOffset();
    switch (t) {
      case "x":
        return qt(r);
      case "xxxx":
      case "xx":
        return Y(r);
      case "xxxxx":
      case "xxx":
      default:
        return Y(r, ":");
    }
  },
  // Timezone (GMT)
  O: function(e, t, n) {
    const r = e.getTimezoneOffset();
    switch (t) {
      case "O":
      case "OO":
      case "OOO":
        return "GMT" + Ht(r, ":");
      case "OOOO":
      default:
        return "GMT" + Y(r, ":");
    }
  },
  // Timezone (specific non-location)
  z: function(e, t, n) {
    const r = e.getTimezoneOffset();
    switch (t) {
      case "z":
      case "zz":
      case "zzz":
        return "GMT" + Ht(r, ":");
      case "zzzz":
      default:
        return "GMT" + Y(r, ":");
    }
  },
  // Seconds timestamp
  t: function(e, t, n) {
    const r = Math.trunc(e.getTime() / 1e3);
    return g(r, t.length);
  },
  // Milliseconds timestamp
  T: function(e, t, n) {
    const r = e.getTime();
    return g(r, t.length);
  }
};
function Ht(e, t = "") {
  const n = e > 0 ? "-" : "+", r = Math.abs(e), a = Math.trunc(r / 60), i = r % 60;
  return i === 0 ? n + String(a) : n + String(a) + t + g(i, 2);
}
function qt(e, t) {
  return e % 60 === 0 ? (e > 0 ? "-" : "+") + g(Math.abs(e) / 60, 2) : Y(e, t);
}
function Y(e, t = "") {
  const n = e > 0 ? "-" : "+", r = Math.abs(e), a = g(Math.trunc(r / 60), 2), i = g(r % 60, 2);
  return n + a + t + i;
}
const Ut = (e, t) => {
  switch (e) {
    case "P":
      return t.date({ width: "short" });
    case "PP":
      return t.date({ width: "medium" });
    case "PPP":
      return t.date({ width: "long" });
    case "PPPP":
    default:
      return t.date({ width: "full" });
  }
}, Gn = (e, t) => {
  switch (e) {
    case "p":
      return t.time({ width: "short" });
    case "pp":
      return t.time({ width: "medium" });
    case "ppp":
      return t.time({ width: "long" });
    case "pppp":
    default:
      return t.time({ width: "full" });
  }
}, kc = (e, t) => {
  const n = e.match(/(P+)(p+)?/) || [], r = n[1], a = n[2];
  if (!a)
    return Ut(e, t);
  let i;
  switch (r) {
    case "P":
      i = t.dateTime({ width: "short" });
      break;
    case "PP":
      i = t.dateTime({ width: "medium" });
      break;
    case "PPP":
      i = t.dateTime({ width: "long" });
      break;
    case "PPPP":
    default:
      i = t.dateTime({ width: "full" });
      break;
  }
  return i.replace("{{date}}", Ut(r, t)).replace("{{time}}", Gn(a, t));
}, Vc = {
  p: Gn,
  P: kc
}, Zc = /^D+$/, Kc = /^Y+$/, Jc = ["D", "DD", "YY", "YYYY"];
function el(e) {
  return Zc.test(e);
}
function tl(e) {
  return Kc.test(e);
}
function nl(e, t, n) {
  const r = rl(e, t, n);
  if (console.warn(r), Jc.includes(e))
    throw new RangeError(r);
}
function rl(e, t, n) {
  const r = e[0] === "Y" ? "years" : "days of the month";
  return `Use \`${e.toLowerCase()}\` instead of \`${e}\` (in \`${t}\`) for formatting ${r} to the input \`${n}\`; see: https://github.com/date-fns/date-fns/blob/master/docs/unicodeTokens.md`;
}
function al(e) {
  return e instanceof Date || typeof e == "object" && Object.prototype.toString.call(e) === "[object Date]";
}
function il(e) {
  if (!al(e) && typeof e != "number")
    return !1;
  const t = A(e);
  return !isNaN(Number(t));
}
const ol = /[yYQqMLwIdDecihHKkms]o|(\w)\1*|''|'(''|[^'])+('|$)|./g, sl = /P+p+|P+|p+|''|'(''|[^'])+('|$)|./g, ul = /^'([^]*?)'?$/, cl = /''/g, ll = /[a-zA-Z]/;
function ze(e, t, n) {
  var l, f, d, m, T, v, w, O;
  const r = Pe(), a = (n == null ? void 0 : n.locale) ?? r.locale ?? Qt, i = (n == null ? void 0 : n.firstWeekContainsDate) ?? ((f = (l = n == null ? void 0 : n.locale) == null ? void 0 : l.options) == null ? void 0 : f.firstWeekContainsDate) ?? r.firstWeekContainsDate ?? ((m = (d = r.locale) == null ? void 0 : d.options) == null ? void 0 : m.firstWeekContainsDate) ?? 1, o = (n == null ? void 0 : n.weekStartsOn) ?? ((v = (T = n == null ? void 0 : n.locale) == null ? void 0 : T.options) == null ? void 0 : v.weekStartsOn) ?? r.weekStartsOn ?? ((O = (w = r.locale) == null ? void 0 : w.options) == null ? void 0 : O.weekStartsOn) ?? 0, s = A(e);
  if (!il(s))
    throw new RangeError("Invalid time value");
  let u = t.match(sl).map((b) => {
    const S = b[0];
    if (S === "p" || S === "P") {
      const E = Vc[S];
      return E(b, a.formatLong);
    }
    return b;
  }).join("").match(ol).map((b) => {
    if (b === "''")
      return { isToken: !1, value: "'" };
    const S = b[0];
    if (S === "'")
      return { isToken: !1, value: fl(b) };
    if (Bt[S])
      return { isToken: !0, value: b };
    if (S.match(ll))
      throw new RangeError(
        "Format string contains an unescaped latin alphabet character `" + S + "`"
      );
    return { isToken: !1, value: b };
  });
  a.localize.preprocessor && (u = a.localize.preprocessor(s, u));
  const c = {
    firstWeekContainsDate: i,
    weekStartsOn: o,
    locale: a
  };
  return u.map((b) => {
    if (!b.isToken)
      return b.value;
    const S = b.value;
    (!(n != null && n.useAdditionalWeekYearTokens) && tl(S) || !(n != null && n.useAdditionalDayOfYearTokens) && el(S)) && nl(S, t, String(e));
    const E = Bt[S[0]];
    return E(s, S, a.localize, c);
  }).join("");
}
function fl(e) {
  const t = e.match(ul);
  return t ? t[1].replace(cl, "'") : e;
}
var Bn = /* @__PURE__ */ ((e) => (e["15seconds"] = "15seconds", e.minute = "minute", e["30minutes"] = "30minutes", e.hourly = "hourly", e.daily = "daily", e.weekly = "weekly", e.monthly = "monthly", e.quarterly = "quarterly", e.yearly = "yearly", e))(Bn || {});
const je = [
  ["15seconds", 15 * 1e3],
  ["minute", 60 * 1e3],
  ["30minutes", 30 * 60 * 1e3],
  ["hourly", 60 * 60 * 1e3],
  ["daily", 24 * 60 * 60 * 1e3],
  ["monthly", 30 * 24 * 60 * 60 * 1e3],
  ["quarterly", 3 * 30 * 24 * 60 * 60 * 1e3],
  ["yearly", 12 * 30 * 24 * 60 * 60 * 1e3]
];
function dl(e, t, n, r, a) {
  const i = t === 0, o = Number(ze(new Date(e), "c")) === 2, s = Number(ze(new Date(e), "q")) === 1, u = t !== 0 ? n[t - 1] : null;
  switch (r) {
    case "15seconds":
      return i || we(e) || oe(e, u) || N(e);
    case "minute":
      return i || we(e) || oe(e, u) || N(e);
    case "30minutes":
      return i || we(e) || oe(e, u) || N(e);
    case "hourly":
      return i || we(e) || oe(e, u) || N(e);
    case "daily":
      return a ? i || o || N(e) : i || oe(e, u) || N(e);
    case "weekly":
      return i || o || N(e);
    case "monthly":
      return i || N(e);
    case "quarterly":
      return i || s;
    case "yearly":
      return !1;
    default:
      throw new Error(`${r} is not a valid time interval.`);
  }
}
function $l(e, t, n, r, a, i) {
  const o = a.showDayName, s = r === "daily" && o ? "weekly" : r, u = new Date(e), c = G(a, "timeIntervalFormats")[s], l = G(c, "primary"), f = G(c, "secondary"), d = dl(e, t, n, r, o);
  let m = d ? l : f;
  r === "15seconds" && u.getMilliseconds() !== 0 && (m = m.replace("pp", "h:mm:ss.SSS a"));
  const T = a.localeObject, { code: v, optionsObject: w } = i, O = w[r].type, b = w[r][d ? "primary" : "secondary"][m];
  if (r === "quarterly" || !b) {
    const E = ze(u, m, { locale: T }).split("").map((R) => {
      let lt = Number(R);
      return R !== " " && !Number.isNaN(lt) ? lt.toLocaleString(v) : R;
    });
    return i[O](u, v, {}, E.join(""));
  } else
    return i[O](u, v, b);
}
function _e(e) {
  const t = new Date(e);
  return {
    M: t.getMonth() + 1,
    // month: 1-12
    d: t.getDate(),
    // day of the month: 1-31
    H: t.getHours(),
    // 24-hour clock: 0-23
    m: t.getMinutes(),
    // minute: 0-59
    s: t.getSeconds()
    // seconds: 0-59
  };
}
function hl(e) {
  if (e)
    return e.slice(1).map((t, n) => t - e[n]);
}
function gl(e) {
  const t = je.reduce((n, [, r], a) => {
    const i = je[n][1], o = Math.abs(i - e), s = Math.abs(r - e);
    return o < s ? n : a;
  }, 0);
  return je[t][0];
}
function Cl(e, t) {
  if (Bn[t])
    return t;
  if (e.length === 1)
    return "15seconds";
  const n = hl(e), r = Zn(n);
  return gl(r);
}
function we(e) {
  const { s: t, m: n, H: r } = _e(e);
  return r === 0 && n === 0 && t === 0;
}
function oe(e, t) {
  const n = _e(e).M, r = _e(t).M;
  return n !== r;
}
function N(e) {
  const { M: t, d: n, s: r, m: a, H: i } = _e(e);
  return t === 1 && n === 1 && i === 0 && a === 0 && r === 0;
}
function Dl(e) {
  return e === void 0 ? "" : e.toISOString();
}
function ml(e) {
  const t = Kn(e, (r) => r.value), n = Jn().domain(t).nice().domain();
  if (n[0] > 0)
    n[0] = 0;
  else if (n[0] === 0 && n[1] === 0)
    return [0, 1];
  return n[0] < 0 && n[1] > 0 && (Math.abs(n[0]) > n[1] ? n[1] = Math.abs(n[0]) : n[0] = -n[1]), n;
}
function Il(e, t) {
  const n = G(t, "gradient", "colors"), r = !qu(n);
  let a = G(t, "pairing", "option");
  const i = ml(e), o = i[0] < 0 && i[1] > 0 ? "diverge" : "mono";
  (a < 1 && a > 4 && o === "mono" || a < 1 && a > 2 && o === "diverge") && (a = 1);
  const s = r ? n : [];
  if (!r) {
    const u = o === "diverge" ? 17 : 11;
    for (let c = 1; c < u + 1; c++)
      s.push(`fill-${o}-${a}-${c}`);
  }
  return er().domain(i).range(s);
}
export {
  ba as $,
  io as A,
  ro as B,
  vl as C,
  Nc as D,
  wl as E,
  uc as F,
  Zu as G,
  Ku as H,
  ut as I,
  tc as J,
  Ju as K,
  xn as L,
  An as M,
  En as N,
  ac as O,
  ic as P,
  oc as Q,
  sc as R,
  $n as S,
  je as T,
  lc as U,
  ye as V,
  fc as W,
  dc as X,
  hc as Y,
  gc as Z,
  cn as _,
  ml as a,
  Ea as a0,
  un as a1,
  sn as a2,
  De as a3,
  Wu as a4,
  go as a5,
  qr as a6,
  he as a7,
  $o as a8,
  ve as a9,
  A as aa,
  q as ab,
  Bc as ac,
  El as ad,
  Al as ae,
  Jt as af,
  st as ag,
  Ca as ah,
  Xs as ai,
  _l as aj,
  xl as ak,
  Sl as al,
  Ol as am,
  Tl as an,
  Pl as ao,
  Yc as ap,
  Ml as aq,
  cc as ar,
  $l as b,
  Cl as c,
  _e as d,
  Lu as e,
  Dl as f,
  Il as g,
  ge as h,
  dl as i,
  Qe as j,
  Ee as k,
  I as l,
  G as m,
  Vu as n,
  Iu as o,
  _ as p,
  Ze as q,
  D as r,
  rc as s,
  qs as t,
  nc as u,
  bl as v,
  Wc as w,
  h as x,
  ec as y,
  qu as z
};
